{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "f037dd59-8d78-4bbc-b1f8-f878afe1f342",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import AlignIO, SeqIO, Seq, pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "e7f5ae9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "min_length_aligned_sequence = 30 #Minimal lenght of sequence which could be an exon\n",
    "extreme_homology = 0.97 #percentage of homology of sequence, treshold #I assume two faulty aligned nucleotides per 100 (98%) and one more nt because sometimes latest nt can move from end of one sequence to beginning next sequence\n",
    "middle_homology = 0.9 #nothing at the moment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "d05490af-896a-4600-92e8-008e685894a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "path11 = './alingment_gen_transkrypt//EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05.aln'\n",
    "path12 = './alingment_gen_transkrypt//EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05.aln'\n",
    "path13 = './alingment_gen_transkrypt//EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05.aln'\n",
    "path21 = './alingment_gen_transkrypt//EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05.aln'\n",
    "path22 = './alingment_gen_transkrypt//EH_OG0030308_t2_g2_EH_PAM1_GP153_OV05.aln'\n",
    "path23 = './alingment_gen_transkrypt//EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05.aln'\n",
    "path31 = './alingment_gen_transkrypt//EL_OG0030307_t1_g1_EL_PAM1_GP153_OV05.aln'\n",
    "path32 = './alingment_gen_transkrypt//EL_OG0030308_t2_g2_EL_PAM1_GP153_OV05.aln'\n",
    "path33 = './alingment_gen_transkrypt//EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05.aln'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "479c14c9-478a-4948-986d-16e849c2e4af",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def cutting_scrap_right(plik, acceptable_gap_length):\n",
    "    \n",
    "    gaps_signs = \"-\" * acceptable_gap_length #maximum length of gaps one by one in sequence\n",
    "    \n",
    "    alignment = AlignIO.read(plik, \"clustal\") #loading data witch \"clustal\" extension \n",
    "    \n",
    "        \n",
    "        \n",
    "    while True:\n",
    "        first_nucleotides_pair = alignment[:, 0]#first nucleotides in both strands as variable\n",
    "        if \"-\" in first_nucleotides_pair: #deleting mismatches at the start #do tej pory wszystko git. program odnajduje znak \"-\" w pierwszej. parze\n",
    "            alignment = alignment[:, 1:]\n",
    "            #print(f\"alignment po obrobce:\\n{alignment}\") #chekpoint\n",
    "            continue\n",
    "    \n",
    "            \n",
    "        \n",
    "        #changing extension\n",
    "        seq1 = alignment[0].seq #from whole variable choose only .seq #upper strand\n",
    "        seq2 = alignment[1].seq #lower strand\n",
    "\n",
    "        \n",
    "        #Counting distance:\n",
    "        first_gaps_position_seq1 = seq1.find(gaps_signs) #getting index of first gap. There is also included gap in genomic sequence - made it intentionally\n",
    "        first_gaps_position_seq2 = seq2.find(gaps_signs)  #this code prints alignment since 1st to certain number (without gaps at the end!)\n",
    "        #print(f\"first gap position in first sequence:{first_gaps_position_seq1}, in 2nd sequence: {first_gaps_position_seq2}\") #checkpoint\n",
    "        \n",
    "        if first_gaps_position_seq2 < 0:\n",
    "            distance_between_first_nt_and_gap = first_gaps_position_seq1\n",
    "        else:\n",
    "            distance_between_first_nt_and_gap = min(first_gaps_position_seq1, first_gaps_position_seq2)\n",
    "\n",
    "        \n",
    "        local_alignment = pairwise2.align.localxx(seq1[:distance_between_first_nt_and_gap], seq2[:distance_between_first_nt_and_gap], \\\n",
    "                                                    one_alignment_only = True)#score_only = True print only score\n",
    "        #I choose function parameters like \"xx\" what means no gap penalty while opening gaps or longering them and no penalties for mismatch. Just pure score of alignment to count homology\n",
    "        #print(\"local_alignment:\", local_alignment) #checkpoint\n",
    "       \n",
    "        try: #obsluga wyjatkow\n",
    "            local_homology_percentage = (local_alignment[0].score / distance_between_first_nt_and_gap) # that sign \"[]\" becouse local_alignment's type is list\n",
    "        except IndexError:\n",
    "            #print(\"index error\", len(seq1))\n",
    "            if len(seq1) < 100:\n",
    "                print(f\"\\n \\n {alignment[0].id} sequence is too short or sequence has not exons. \\n\")\n",
    "                break\n",
    "            continue\n",
    "        \n",
    "        \n",
    "        if local_homology_percentage < extreme_homology:\n",
    "            #print(f\"\\n Too low score of homology. \\n alignment: {alignment[:, :50]} \\n score: {local_alignment[0].score}, distance: {distance_between_first_nt_and_gap}, percent: {round(local_homology_percentage, 2)*100}%, \\n {local_alignment}\")\n",
    "            alignment = alignment[:, distance_between_first_nt_and_gap:] \n",
    "            \n",
    "        else:\n",
    "            if distance_between_first_nt_and_gap < min_length_aligned_sequence:\n",
    "                #print(f\"\\n alignment is too short. \\n distance between sequences {distance_between_first_nt_and_gap}, \\n alignment: {alignment}, \\n score: {local_alignment[0].score}, distance: {distance_between_first_nt_and_gap}, procentowo: {round(local_homology_percentage, 2)*100}%, \\n {local_alignment}\")\n",
    "                alignment = alignment[:, distance_between_first_nt_and_gap:]\n",
    "            else:\n",
    "                #print(f\"Sequence {alignment[0].id} cut properly from LEFT side\")\n",
    "                #print(f\"\"\"\\n \\n \\n victory lewo! \\n distance between sequences: {distance_between_first_nt_and_gap},\n",
    "                #alignment: {alignment[:, :50]}, \n",
    "                #score: {local_alignment[0].score}, dystans: {distance_between_first_nt_and_gap}, \n",
    "                #percent: {round(local_homology_percentage, 4)*100}%, \\n \\n local alignment to confirm: \\n{local_alignment}\"\"\")\n",
    "                break\n",
    "        \n",
    "\n",
    "    \n",
    "######################### TERAZ OD PRAWEJ DO LEWEJ #######################  \n",
    "\n",
    "    \n",
    "    while True:\n",
    "        last_nucleotides_pair = alignment[:, -1]#last nucleotides in both strands as variable\n",
    "        if \"-\" in last_nucleotides_pair: #deleting mismatches at the end\n",
    "            alignment = alignment[:, :-1]\n",
    "            #print(f\"alignment check:\\n{alignment[:, -50:-1]}\") checkpoint\n",
    "            continue\n",
    "                \n",
    "            \n",
    "#tu jest problem - funkcja usuwa o jedną pare nukleotydow za duzo, mimo ze nie są ona gap \"-\". - zachodzi jedna dodatkowa iteracja Na razie pomijam\n",
    "        #jak nic nie podziała to trzeba bedzie zmierzyc odleglosc od ostatniego nukleotydu do pierwszego (szerokosc gapow) i usunac to w petli for\n",
    "        \n",
    "        #changing extension\n",
    "        seq1 = alignment[0].seq #from whole file choose only .seq #upper strand\n",
    "        seq2 = alignment[1].seq #lower strand\n",
    "        \n",
    "       \n",
    "        last_gaps_position_seq1 = seq1.rfind(gaps_signs)  #Counting distance: #getting index of first gap. There is also included gap in genomic sequence - made it intentionally\n",
    "        last_gaps_position_seq1 = last_gaps_position_seq1 + acceptable_gap_length\n",
    "        \n",
    "        last_gaps_position_seq2 = seq2.rfind(gaps_signs)  #this code prints alignment since 1st to certain number (without gaps at the end!)\n",
    "        last_gaps_position_seq2 = last_gaps_position_seq2 + acceptable_gap_length #now we have to add length of gap to index, because python starts counting from 0 nt, and we just change direction of counting.\n",
    "        #print(\"seq1:\", last_gaps_position_seq1, \"seq2:\", last_gaps_position_seq2) checkpoint\n",
    "        \n",
    "        \n",
    "        if last_gaps_position_seq2 < 0: #it means, if there is no gaps in second sequention, gap position is gap position in seq1. Additional, rfind() function return \"-1\" if it won't find any gaps in seq\n",
    "            distance_between_first_nt_and_gap = (len(seq1) - last_gaps_position_seq1)\n",
    "            distance_between_first_nt_and_gap = distance_between_first_nt_and_gap - 1 #correction to counting from 0th index\n",
    "            #print(\"pierwszy\", distance_between_first_nt_and_gap)\n",
    "        else:\n",
    "            distance_between_first_nt_and_gap = (len(seq1) - max(last_gaps_position_seq1, last_gaps_position_seq2))#tutaj napisalem taki kod, ktory wybiera najmniejsza odleglosc miedzy pierwszym nukleotydem a najblizsza podwojna przerwa. Ale to moze uwzglednic rowniez przerwe w sekwencji genowej - tak zostawiam\n",
    "            #print(\"drugi\", distance_between_first_nt_and_gap)\n",
    "            \n",
    "            \n",
    "            local_alignment = pairwise2.align.localxx(seq1[-distance_between_first_nt_and_gap:], seq2[-distance_between_first_nt_and_gap:], \\\n",
    "                                                    one_alignment_only = True) #score_only = True print only score #I choose function parameters like \"xx\" what means no gap penalty while opening gaps or longering them and no penalties for mismatch. Just pure score of alignment\n",
    "        #print(\"local_alignment:\", local_alignment) #checkpoint\n",
    "        #mam wrazenie ze python nie chce mi lapac ostatniego nukleotydu w komendzie: [liczba:] - zrobil alignment dla 180nt gdzie dlugosc gapu to 180\n",
    "\n",
    "        \n",
    "\n",
    "        try: #obsluga wyjatkow\n",
    "            local_homology_percentage = (local_alignment[0].score / distance_between_first_nt_and_gap) #that sign \"[]\" becouse local_alignment's type is list\n",
    "        except IndexError:\n",
    "            if len(seq1) < 100:\n",
    "                print(f\"\\n \\n {alignment[0].id} sequence is too short or sequence has not exons. \\n\")\n",
    "                break\n",
    "            continue\n",
    "            #niestety ten warunek jakoś nie działa na sekwencje 32\n",
    "\n",
    "        \n",
    "        if local_homology_percentage < extreme_homology:\n",
    "            #print(f\"\\n Alignment ma zbyt niska homologie. \\n alignment: {alignment[:, -50:]} \\n score: {local_alignment[0].score}, \\dystans: {distance_between_first_nt_and_gap}, procentowo: {round(local_homology_percentage, 2)*100}%, \\n {local_alignment}\")\n",
    "            alignment = alignment[:, :-distance_between_first_nt_and_gap] \n",
    "            \n",
    "        else:\n",
    "            if distance_between_first_nt_and_gap < min_length_aligned_sequence:\n",
    "                #print(f\"\\n alignment jest za krotki. \\n dystans miedzy sekwencjami {distance_between_first_nt_and_gap}, \\n alignment: {alignment}, \\n score: {local_alignment[0].score}, dystans: {distance_between_first_nt_and_gap}, procentowo: {round(local_homology_percentage, 2)*100}%, \\n {local_alignment}\")\n",
    "                alignment = alignment[:, :-distance_between_first_nt_and_gap]\n",
    "            else:\n",
    "                #print(f\"Sequence: {alignment[0].id} cut properly from RIGHT side\")\n",
    "                #print(f\"\"\"\\n \\n \\n victory prawo! \\n distance between sequences {distance_between_first_nt_and_gap},\n",
    "                #alignment: {alignment[:, -50:]}, \n",
    "                #score: {local_alignment[0].score}, distance: {distance_between_first_nt_and_gap}, \n",
    "                #percent: {round(local_homology_percentage, 4)*100}%, \\n \\n local alignment to confirm: \\n{local_alignment}\"\"\")\n",
    "                break\n",
    "\n",
    "    index = 0\n",
    "    id = []\n",
    "    is_it_intron = []\n",
    "    class_of_exon = []\n",
    "    exon_or_intron_number = []\n",
    "    percent_of_homology_for_exons = []\n",
    "    length = []\n",
    "    first_10_nt = []\n",
    "    last_10_nt = []\n",
    "    sequence = []\n",
    "    path = []\n",
    "    \n",
    "    while True: #chyba mimo wszystko zostajemy przy while true bo trudno jest mi wyciagnac liczbe plikow ktore beda lacznie wprowadzane. Tak jest chyba łatwiej\n",
    "        #problem taki wynikajacy ze specyfikacji programu jest taki, że outputowa tabelka bedzie tylko dla jednego organizmu. Potem mozna je jakos polaczyc w calosc\n",
    "        try: #obsluga wyjatkow\n",
    "            local_homology_percentage = (local_alignment[0].score / distance_between_first_nt_and_gap) #that sign \"[]\" becouse local_alignment's type is list\n",
    "        except IndexError:\n",
    "            if len(seq1) < 100:\n",
    "                print(f\"\\n \\n {alignment[0].id} sequence is too short or sequence has not exons. \\n\")\n",
    "                break\n",
    "            continue\n",
    "        #chwila chwila, czy w tym momencie jest nam wogole potrzebna ta petla while true? bo ona tylko opetla jedna petle ktora sama sie wygasza po spelnieniu warunków\n",
    "        nucleotides = [\"a\", \"t\", \"g\", \"c\", \"A\", \"T\", \"G\", \"C\"]\n",
    "\n",
    "        #plan jest taki: napierdalamy z karabinkow ile wzlezie\n",
    "        #plan jest taki: szukamy 1nt eksonu, zrobimy warunek ze jesli wynikiem alignment_of_exon bedzie NIC (tak jak tearz) to puszczamy alignment na intronie dopisujac tylko index i sekwencje druga\n",
    "        #EKSON INTRON EXON INTRON EXON INTRON\n",
    "        #dla pierwszego eksonu robimy alignment i go zapisujemy do df\n",
    "        #ucinamy ten ekosn tak jak wczesniej smieci (wyzej)\n",
    "        #robimy alignment dla intronu i go ucinamy\n",
    "        #mozemy to zapetlic az do momentu gdy dlugosc seq1 bedzie = 0\n",
    "\n",
    "        #ewentualnie nie trzeba wcale tego przycinać i eliminowac bo mozna przesuwac obszar wyszukiwania przez funckje find() podajac jej miejsce startu i stopu. Ale w sumie to na jedno wychodzi. Zalezy co jest szybsze dla komputera\n",
    "        #string.find(value, start, end) #ale musielibysmy i tak tworzyc dwie odnogi zeby wklepac jej raz wzor szukania \"--\"        a drugim razem ktorys z nukleotydow\n",
    "        \n",
    "\n",
    "        if len(seq1) > 1:\n",
    "            #print(\"Uruchomienie petli\")\n",
    "            #sequence1 = alignment[0].seq #musialem dodac te dwie linijki bo ucinanie tego intronu nie wychodzilo poza ta petle. Nie uzylem wczesniej napisanej funkcji zeby uniknac bledu jakiegos\n",
    "            #sequence2 = alignment[1].seq\n",
    "            if seq1[0] != gaps_signs: #mining data from exons   \n",
    "                #print(\"uruchomienie analizy eksonow\")\n",
    "                index += 1\n",
    "                last_nt_exon_index = seq1.find(gaps_signs) #zatrzymuje sie gdy znajdzie gapa - dwa gapy\n",
    "                \n",
    "                if last_nt_exon_index == -1: #it means if it is last exon\n",
    "                    last_nt_exon_index = len(seq1)\n",
    "                    \n",
    "                \n",
    "                alignment_of_exon = pairwise2.align.localxx(seq1[:last_nt_exon_index], seq2[:last_nt_exon_index], one_alignment_only = True) #zdaje sobie sprawe ze score = 94 gdy dlugosc = 96, czyli homologia < 97%. Ale to dlatego ze funkcja wprowadza swoje wlasne gapy, ktore i tak nie maja znaczenia bo liczy sie tylko dopasowanie czyli poziom identycznych nukleotydow\n",
    "                #alignment = alignment[:, last_nt_exon_index:] #nie zadziala bo tutaj elementem przycinanym powinna być seq1\n",
    "\n",
    "                #counting percent of homology\n",
    "                percent_of_homology = alignment_of_exon[0].score / last_nt_exon_index\n",
    "                #print(percent_of_homology)\n",
    "                \n",
    "                #print(\"zakonczenie analizy eksonow\")\n",
    "                #print(\"score alignmentu = \", alignment_of_exon[0].score)\n",
    "                #print(\"seq1 = \", seq1)\n",
    "                #print(\"alignment = \", alignment_of_exon[0])\n",
    "                \n",
    "\n",
    "                #TWORZENIE TABELKI\n",
    "                id.append(alignment[0].id)\n",
    "                exon_or_intron_number.append(index)\n",
    "                is_it_intron.append(False)\n",
    "                percent_of_homology_for_exons.append(round(percent_of_homology * 100, 2))#tylko dla exonow\n",
    "                length.append(last_nt_exon_index)\n",
    "                first_10_nt.append(str(seq1[:10]))\n",
    "                last_10_nt.append(str(seq1[last_nt_exon_index-10 : last_nt_exon_index])) #to jest 10 ostatnich nukleotydow sekwencji calej, nie alignmentu\n",
    "                sequence.append(str(seq1[:last_nt_exon_index])) #odblokuj ta opcje w przyszlosci! #tylko dla exonow!\n",
    "                path.append(str(plik))\n",
    "                \n",
    "                if percent_of_homology >= 0.9:\n",
    "                    class_of_exon.append(\"1\")\n",
    "                elif percent_of_homology < 0.9 and percent_of_homology > 0:\n",
    "                    class_of_exon.append(\"2\")\n",
    "                else:\n",
    "                    class_of_exon.append(\"3\")\n",
    "                \n",
    "                seq1 = seq1[last_nt_exon_index:]\n",
    "                seq2 = seq2[last_nt_exon_index:] #BINGO tym skracamy seq1\n",
    "                #print(\"zakonczenie analizy eksonow\")\n",
    "\n",
    "                \n",
    "\n",
    "            if len(seq1) > 1 and seq1[0] == \"-\":\n",
    "                index += 1\n",
    "                #print(\"uruchomienie analizy intronow\")\n",
    "                last_nt_intron_index = None #searching for first nt in sequence (a, t, g, or c)\n",
    "                for nt in nucleotides:\n",
    "                    nt_index = seq1.find(nt)\n",
    "                    if nt_index != -1:\n",
    "                        if last_nt_intron_index is None or last_nt_intron_index > nt_index:\n",
    "                            last_nt_intron_index = nt_index -1\n",
    "                            break #bede musial to skontrolowac, cyz przypadkiem ten break nie hamuje calej funkcji przed znalezieniem najmniejszego indeksu\n",
    "                    else:\n",
    "                        break\n",
    "                if last_nt_intron_index == seq2[-1]:\n",
    "                    print(\"Cutting alignment from the end was faulty!\")\n",
    "                \n",
    "                \n",
    "                id.append(alignment[0].id)\n",
    "                exon_or_intron_number.append(index)\n",
    "                is_it_intron.append(True) #czy te cudzyslowy sa potrzebne? czy to bedzie traktowane jako string czy jako true or false?\n",
    "                percent_of_homology_for_exons.append(None)#tylko dla exonow\n",
    "                length.append(last_nt_intron_index)\n",
    "                first_10_nt.append(str(seq2[:10]))\n",
    "                last_10_nt.append(str(seq2[last_nt_intron_index-10 : last_nt_intron_index]))\n",
    "                sequence.append(str(seq2[:last_nt_intron_index])) #odblokuj ta opcje w przyszlosci! #tylko dla exonow!\n",
    "                path.append(str(plik))\n",
    "                class_of_exon.append(None)\n",
    "\n",
    "                seq1 = seq1[last_nt_intron_index:]\n",
    "                seq2 = seq2[last_nt_intron_index:]\n",
    "                #print(\"zakonczenie analizy intronow\")\n",
    "            else:\n",
    "                break\n",
    "                \n",
    "\n",
    "                \n",
    "    dictionary = {\"ID\":id, \"exon_or_intron_number\":exon_or_intron_number, \"is_it_intron\":is_it_intron, \"class_of_exon\":class_of_exon, \"percent_of_homology_for_exons\":percent_of_homology_for_exons, \"length\":length, \"first_10_nt\" : first_10_nt, \"last_10_nt\" : last_10_nt, \"path\":path, \"sequence\" : sequence}\n",
    "    #print(dictionary)        \n",
    "    data_frame = pd.DataFrame(dictionary) #ramke mamy dobrze zrobiona, niestety ona na razie pobiera dane z tylko jednego pliku wejsciowego i z tylko pierwszego eksonu\n",
    "    print()\n",
    "    print()\n",
    "    print(\"\\n\\n\\n\\n\\n\", data_frame, \"\\n\\n\\n\\n\\n\")\n",
    "\n",
    "        \n",
    "        \n",
    "        #fragment zapisujacy wartosci do ramki\n",
    "#uładnij kod, usun zbedne komentarze, dodaj te niezbedne po angielsku \n",
    "\n",
    "#na razie ta funkcja dziala tak: jedna sciezka, jedna ramka. Musze zmienic to tak by wiele sciezek = jedna ramka\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "c4059512-cac4-41e9-86b2-22c0f9f34abd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#cutting_scrap_right(path11, 2)\n",
    "#cutting_scrap_right(path12, 2)\n",
    "#cutting_scrap_right(path13, 2)\n",
    "#cutting_scrap_right(path21, 2)\n",
    "#cutting_scrap_right(path22, 2)\n",
    "#cutting_scrap_right(path23, 2)\n",
    "#cutting_scrap_right(path31, 2)\n",
    "#cutting_scrap_right(path32, 2)\n",
    "#cutting_scrap_right(path33, 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "1d0dbc7a-160a-4ae3-96a7-1cbd4df8f3c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                       ID  exon_or_intron_number  is_it_intron  \\\n",
      "0  EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                      1         False   \n",
      "1  EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                      2          True   \n",
      "2  EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                      3         False   \n",
      "3  EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                      4          True   \n",
      "4  EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                      5         False   \n",
      "5  EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                      6          True   \n",
      "6  EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                      7         False   \n",
      "\n",
      "  class_of_exon  percent_of_homology_for_exons  length first_10_nt  \\\n",
      "0             1                          97.92      96  ctacacgtac   \n",
      "1          None                            NaN     102  gcccgccgcc   \n",
      "2             1                         100.00     145  gatcgagctg   \n",
      "3          None                            NaN     237  gaccgggcgg   \n",
      "4             1                          98.63      73  -atgtcgccg   \n",
      "5          None                            NaN     399  cggccagggc   \n",
      "6             1                          97.89     190  cagctgcagg   \n",
      "\n",
      "   last_10_nt                                               path  \\\n",
      "0  ggctgctcct  ./alingment_gen_transkrypt/EG_OG0030307_t1_g1_...   \n",
      "1  cggcgggggc  ./alingment_gen_transkrypt/EG_OG0030307_t1_g1_...   \n",
      "2  ccgctgcttt  ./alingment_gen_transkrypt/EG_OG0030307_t1_g1_...   \n",
      "3  ggcccgcggt  ./alingment_gen_transkrypt/EG_OG0030307_t1_g1_...   \n",
      "4  caccgtggca  ./alingment_gen_transkrypt/EG_OG0030307_t1_g1_...   \n",
      "5  ggcctggggg  ./alingment_gen_transkrypt/EG_OG0030307_t1_g1_...   \n",
      "6  agccccccgg  ./alingment_gen_transkrypt/EG_OG0030307_t1_g1_...   \n",
      "\n",
      "                                            sequence  \n",
      "0  ctacacgtacaagcacgtgaacgtcatggcgccggaggtccaaagc...  \n",
      "1  gcccgccgcccggggttcggtgcccccggctgtcaggccgttgggt...  \n",
      "2  gatcgagctgctgaccggccagccggcgttccccgtctcggaggac...  \n",
      "3  gaccgggcggcctcaccatggcagtgccacatcctgggccaacaga...  \n",
      "4  -atgtcgccggacgccctctcgctggccctccgctgcctcgccaag...  \n",
      "5  cggccagggccgggcaggacaccccacgactcctgcagttcagggg...  \n",
      "6  cagctgcaggcggacccgttcttcgcggacgtggccgtgaagcgcc...   \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                        ID  exon_or_intron_number  \\\n",
      "0   EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                      1   \n",
      "1   EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                      2   \n",
      "2   EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                      3   \n",
      "3   EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                      4   \n",
      "4   EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                      5   \n",
      "5   EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                      6   \n",
      "6   EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                      7   \n",
      "7   EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                      8   \n",
      "8   EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                      9   \n",
      "9   EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                     10   \n",
      "10  EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                     11   \n",
      "11  EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                     12   \n",
      "12  EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                     13   \n",
      "13  EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                     14   \n",
      "14  EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                     15   \n",
      "15  EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                     16   \n",
      "16  EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                     17   \n",
      "17  EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                     18   \n",
      "18  EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                     19   \n",
      "19  EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                     20   \n",
      "20  EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                     21   \n",
      "21  EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                     22   \n",
      "22  EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                     23   \n",
      "\n",
      "    is_it_intron class_of_exon  percent_of_homology_for_exons  length  \\\n",
      "0          False             1                          97.78      90   \n",
      "1           True          None                            NaN     841   \n",
      "2          False             1                          93.62      47   \n",
      "3           True          None                            NaN     235   \n",
      "4          False             1                          99.44     179   \n",
      "5           True          None                            NaN     344   \n",
      "6          False             1                         100.00     153   \n",
      "7           True          None                            NaN     624   \n",
      "8          False             1                         100.00      84   \n",
      "9           True          None                            NaN     313   \n",
      "10         False             1                         100.00      66   \n",
      "11          True          None                            NaN    2220   \n",
      "12         False             1                         100.00     140   \n",
      "13          True          None                            NaN     558   \n",
      "14         False             1                         100.00      80   \n",
      "15          True          None                            NaN    1604   \n",
      "16         False             1                         100.00      96   \n",
      "17          True          None                            NaN     227   \n",
      "18         False             1                         100.00      76   \n",
      "19          True          None                            NaN     309   \n",
      "20         False             1                         100.00      38   \n",
      "21          True          None                            NaN     760   \n",
      "22         False             1                          99.01     202   \n",
      "\n",
      "   first_10_nt  last_10_nt                                               path  \\\n",
      "0   ttcttctcgg  tctgctgcaa  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "1   ggggccaggc  ccctcttcct  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "2   gaaggcggag  cacctccccg  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "3   accgcccaga  gcatgtctgc  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "4   -accggctgc  cccaggtccc  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "5   cggccgaggc  tctcgggggc  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "6   gaggtcatct  gatcggccag  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "7   cagccaatgg  cgattgccac  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "8   caattgcggc  ggattctgaa  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "9   cacccaagtg  cacttgttgc  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "10  catctgatga  tgtctgcccc  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "11  gagcaccccc  cgctggcggg  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "12  gacatcgagc  gggccggcca  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "13  cgcctggggc  ctcccccggt  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "14  gaccatcacc  gtctcgcgct  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "15  ccaccctggg  gcccagcggc  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "16  caagctgcgc  agtacggaga  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "17  gagcagggga  ccctggaggc  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "18  gagcttcggg  cgtggagggc  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "19  ctgggcaggg  gctggggctg  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "20  gacatgtacg  agccggaggc  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "21  ggccgtgcca  gggtgctctc  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "22  caggcccccg  tataaccagg  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "\n",
      "                                             sequence  \n",
      "0   ttcttctcggagcgccccatgacgcccatcagccacgccgtgttcc...  \n",
      "1   ggggccaggctacgcgggaggggcgtttacatccgaagcacagtat...  \n",
      "2     gaaggcggagctgcccatccgcttggcccacgccgcgcacctccccg  \n",
      "3   accgcccagacattcacctagtgtgcccacaaaacaatgttgcatc...  \n",
      "4   -accggctgccccggggcctgggcggcctggcagcgttccagcagg...  \n",
      "5   cggccgaggctcccatgccccatggtgggtgtgggacagggggcga...  \n",
      "6   gaggtcatcttccagggctgccagcagctccgcgttcagctgcagc...  \n",
      "7   cagccaatggttcgtgtacttccccctacattgaagccccacagtg...  \n",
      "8   caattgcggctccatgcacagtacgtccgaggggagagcccctccg...  \n",
      "9   cacccaagtgccatctgcacaccattaaccaggaaggctctgcttc...  \n",
      "10  catctgatgaaggcgtccatcgcaactgcgaagcgcatggccatcg...  \n",
      "11  gagcaccccccctgtgacactccaacaggtggcttctttaaattca...  \n",
      "12  gacatcgagctcacctgcaagtgtggcacgcccttcatccagatcg...  \n",
      "13  cgcctggggcttggtctaagcaagccaggatccggatgaaagccag...  \n",
      "14  gaccatcaccgtggcggacgccgggggcaacgaggacctgtgcatc...  \n",
      "15  ccaccctgggctcgttggctggcccctggccacaagcaaggttggg...  \n",
      "16  caagctgcgctacctgtggagctacctgcacgatgatgctcgcctc...  \n",
      "17  gagcaggggatggtcttatgccaatttaatattgtcccgtgcgtcg...  \n",
      "18  gagcttcgggctgcccctcgcccgcctgcgggcccggctgttcggc...  \n",
      "19  ctgggcagggctgaccgggacgcccccaaaacaccccccagggtgg...  \n",
      "20             gacatgtacgcgtacatcccgaagttcaagccggaggc  \n",
      "21  ggccgtgccacagcttcatggcatcagggtttctggcactggtgct...  \n",
      "22  caggcccccgttcccgaattgatgtgcgcttttgggcgcattcgca...   \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                        ID  exon_or_intron_number  \\\n",
      "0   EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                      1   \n",
      "1   EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                      2   \n",
      "2   EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                      3   \n",
      "3   EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                      4   \n",
      "4   EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                      5   \n",
      "5   EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                      6   \n",
      "6   EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                      7   \n",
      "7   EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                      8   \n",
      "8   EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                      9   \n",
      "9   EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                     10   \n",
      "10  EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                     11   \n",
      "11  EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                     12   \n",
      "12  EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                     13   \n",
      "13  EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                     14   \n",
      "14  EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                     15   \n",
      "15  EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                     16   \n",
      "16  EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                     17   \n",
      "17  EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                     18   \n",
      "18  EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                     19   \n",
      "\n",
      "    is_it_intron class_of_exon  percent_of_homology_for_exons  length  \\\n",
      "0          False             1                         100.00     169   \n",
      "1           True          None                            NaN     631   \n",
      "2          False             1                         100.00     107   \n",
      "3           True          None                            NaN     178   \n",
      "4          False             1                         100.00     107   \n",
      "5           True          None                            NaN     315   \n",
      "6          False             1                         100.00      55   \n",
      "7           True          None                            NaN     571   \n",
      "8          False             1                         100.00     149   \n",
      "9           True          None                            NaN     407   \n",
      "10         False             1                         100.00      89   \n",
      "11          True          None                            NaN     231   \n",
      "12         False             1                          99.25     134   \n",
      "13          True          None                            NaN    1813   \n",
      "14         False             1                          98.46      65   \n",
      "15          True          None                            NaN     157   \n",
      "16         False             1                         100.00      38   \n",
      "17          True          None                            NaN     436   \n",
      "18         False             1                         100.00     144   \n",
      "\n",
      "   first_10_nt  last_10_nt                                               path  \\\n",
      "0   gggcgctgtt  gggcccgccc  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "1   cgccccttgg  cgtcctgccg  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "2   caacccattc  cttgaggaac  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "3   ggtatcggcc  gcgcagggtc  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "4   caccacgccc  tacagggggc  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "5   gtccaggctt  ggtgtctttg  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "6   gagaccctgc  gtgcgagtct  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "7   gctgggtgcc  tggccggctg  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "8   tacagcgtgc  ccatctacac  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "9   gtgtgccccg  gtcgcacagg  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "10  tacaagcacg  ggctgctcct  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "11  gtccaggcct  gccctggcgg  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "12  -aattgagct  ccggtgctct  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "13  ggccgggctg  cgtccggggc  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "14  -atgacgccg  cgccggccca  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "15  ccgtggccag  gcccgtggcc  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "16  cagctgctgg  cccctgcggt  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "17  gcccaggctg  gcctgctggg  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "18  gaagcggcag  cgccgagccc  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "\n",
      "                                             sequence  \n",
      "0   gggcgctgttctgtgccacacccataataccaatggcggaggagga...  \n",
      "1   cgccccttgggcgttgcctccgaccccccctccccaaatgttagca...  \n",
      "2   caacccattccatctcgggcctcggcggaaaaatggcgatttcctc...  \n",
      "3   ggtatcggccctcccaccccaatgcggtatttttacttttcctctc...  \n",
      "4   caccacgcccggggcatgacccacgaccacctcgtgccgtatgtgg...  \n",
      "5   gtccaggctttgacgcaaggaggcttgtagctaaacaggtggaggg...  \n",
      "6   gagaccctgctgcggaaacgcaggcgcgggttcgatgagatgcagg...  \n",
      "7   gctgggtgccgaccgcactccacatgccacggttgtcatgtctatt...  \n",
      "8   tacagcgtgctgaagggcctgcagcacctccaccagcacggctgcg...  \n",
      "9   gtgtgccccgccccacatccccccgacccaggcaaccggcccccta...  \n",
      "10  tacaagcacgtcaatgtgatggcgccagaggtccagagcggctcgg...  \n",
      "11  gtccaggcctatcgtggtttatatgggacagaccgcgctcgtgcct...  \n",
      "12  -aattgagctcctgactggacagcccgccttccccgtctccgagga...  \n",
      "13  ggccgggctgcgctctggcccccactgtattgagggaggataggga...  \n",
      "14  -atgacgccggaagctctcgccctcgccctgcgctgcctcgccccg...  \n",
      "15  ccgtggccagccttctttttaagggctgctggaccccccccccctg...  \n",
      "16             cagctgctggcggacgacttctttgccgcccctgcggt  \n",
      "17  gcccaggctgaagttggcattgcattggacctcaatcatccccccc...  \n",
      "18  gaagcggcaggcgctgctcggcctgaagcgggcgttgacggaccgc...   \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                        ID  exon_or_intron_number  \\\n",
      "0   EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                      1   \n",
      "1   EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                      2   \n",
      "2   EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                      3   \n",
      "3   EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                      4   \n",
      "4   EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                      5   \n",
      "5   EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                      6   \n",
      "6   EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                      7   \n",
      "7   EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                      8   \n",
      "8   EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                      9   \n",
      "9   EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                     10   \n",
      "10  EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                     11   \n",
      "11  EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                     12   \n",
      "12  EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                     13   \n",
      "13  EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                     14   \n",
      "14  EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                     15   \n",
      "15  EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                     16   \n",
      "16  EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                     17   \n",
      "\n",
      "    is_it_intron class_of_exon  percent_of_homology_for_exons  length  \\\n",
      "0          False             1                         100.00     227   \n",
      "1           True          None                            NaN     440   \n",
      "2          False             1                         100.00     226   \n",
      "3           True          None                            NaN     316   \n",
      "4          False             1                         100.00     154   \n",
      "5           True          None                            NaN      82   \n",
      "6          False             1                          98.81      84   \n",
      "7           True          None                            NaN     964   \n",
      "8          False             1                         100.00      66   \n",
      "9           True          None                            NaN     292   \n",
      "10         False             1                         100.00     140   \n",
      "11          True          None                            NaN     359   \n",
      "12         False             1                         100.00      83   \n",
      "13          True          None                            NaN     751   \n",
      "14         False             1                         100.00     219   \n",
      "15          True          None                            NaN     779   \n",
      "16         False             1                          99.41     170   \n",
      "\n",
      "   first_10_nt  last_10_nt                                               path  \\\n",
      "0   accgtcaagt  tgttgcaggg  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "1   gcccaggccc  ccctcttcct  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "2   caaggcggag  gcacgtccct  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "3   ggacgatgcc  cctcgccatc  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "4   gaggtgattt  gtcggtcagc  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "5   agccagtgcc  cctctgccaa  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "6   -agctgcagc  ggattctgaa  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "7   cgcccaagtg  catttgcagt  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "8   cacctgatga  cgtctgtcct  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "9   caccagcgtt  cgctggctgg  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "10  gacattgaac  aggaaggcaa  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "11  tgcctgggtg  ctgttcctcc  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "12  gatcactgtg  atctcccggt  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "13  ccaacctgag  gcccagtggc  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "14  caagttgcgc  agccagagga  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "15  gtctgtgctg  gtgctttctg  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "16  gacccgcgtt  agtggtgcac  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "\n",
      "                                             sequence  \n",
      "0   accgtcaagtgcctgtcaaccgattccgttggcaaccgccttgagc...  \n",
      "1   gcccaggccccggcctgccattccaggcccggtcatgcttgacttg...  \n",
      "2   caaggcggagctgcccatccgcctggcccactccgcccatctgctc...  \n",
      "3   ggacgatgcccctcgtttcccctccccaacccccccacctggccgg...  \n",
      "4   gaggtgattttccagggctgcaagcaacttcgcgctcagctgcacc...  \n",
      "5   agccagtgcctggccactcgatggtgggcccttcttttcctcccgc...  \n",
      "6   -agctgcagctccacgcacagtacatcggaggcgaaagcccctcag...  \n",
      "7   cgcccaagtgcccaagcctccgaagaaaggaaatgacagcaaagaa...  \n",
      "8   cacctgatgaaggcatccattgcaacctgcaagcggatggccatcg...  \n",
      "9   caccagcgttgcttttgcattgggcacattccgagttcgcagcttt...  \n",
      "10  gacattgaactcacctgcaactgtgggacaccgctggtgcaaatcg...  \n",
      "11  tgcctgggtgcgggtttcagcatttgcacaaaagggttgtggaccc...  \n",
      "12  gatcactgtgacagtggcggatgccggcaccaatgaggacttgtgc...  \n",
      "13  ccaacctgagcccactgactgcaccggaaaatgatgcagtgtcccc...  \n",
      "14  caagttgcgccagctgtggagctatttgcacgacgacgctcacctt...  \n",
      "15  gtctgtgctgcagcttgctggtcagcggggggggggggcaggcagg...  \n",
      "16  gacccgcgtttcaaaagtgacctcctattcgtgtcccctgctcgct...   \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " \n",
      " EL_OG0030308_t2_g2_EL_PAM1_GP153_OV05 sequence is too short or sequence has not exons. \n",
      "\n",
      "\n",
      " \n",
      " EL_OG0030308_t2_g2_EL_PAM1_GP153_OV05 sequence is too short or sequence has not exons. \n",
      "\n",
      "\n",
      " \n",
      " EL_OG0030308_t2_g2_EL_PAM1_GP153_OV05 sequence is too short or sequence has not exons. \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " Empty DataFrame\n",
      "Columns: [ID, exon_or_intron_number, is_it_intron, class_of_exon, percent_of_homology_for_exons, length, first_10_nt, last_10_nt, path, sequence]\n",
      "Index: [] \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                        ID  exon_or_intron_number  \\\n",
      "0   EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                      1   \n",
      "1   EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                      2   \n",
      "2   EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                      3   \n",
      "3   EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                      4   \n",
      "4   EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                      5   \n",
      "5   EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                      6   \n",
      "6   EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                      7   \n",
      "7   EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                      8   \n",
      "8   EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                      9   \n",
      "9   EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     10   \n",
      "10  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     11   \n",
      "11  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     12   \n",
      "12  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     13   \n",
      "13  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     14   \n",
      "14  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     15   \n",
      "15  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     16   \n",
      "16  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     17   \n",
      "17  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     18   \n",
      "18  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     19   \n",
      "19  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     20   \n",
      "20  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     21   \n",
      "21  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     22   \n",
      "22  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     23   \n",
      "23  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     24   \n",
      "24  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     25   \n",
      "25  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     26   \n",
      "26  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     27   \n",
      "27  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     28   \n",
      "28  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     29   \n",
      "29  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     30   \n",
      "30  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     31   \n",
      "31  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     32   \n",
      "32  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     33   \n",
      "33  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     34   \n",
      "34  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     35   \n",
      "35  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     36   \n",
      "36  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     37   \n",
      "37  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     38   \n",
      "38  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     39   \n",
      "39  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     40   \n",
      "40  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     41   \n",
      "41  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     42   \n",
      "42  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     43   \n",
      "43  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     44   \n",
      "44  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     45   \n",
      "45  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     46   \n",
      "46  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     47   \n",
      "47  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     48   \n",
      "48  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     49   \n",
      "49  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     50   \n",
      "50  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     51   \n",
      "51  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     52   \n",
      "52  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     53   \n",
      "53  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     54   \n",
      "54  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     55   \n",
      "55  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     56   \n",
      "56  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     57   \n",
      "\n",
      "    is_it_intron class_of_exon  percent_of_homology_for_exons  length  \\\n",
      "0          False             1                         100.00     150   \n",
      "1           True          None                            NaN     196   \n",
      "2          False             1                         100.00      56   \n",
      "3           True          None                            NaN      50   \n",
      "4          False             1                         100.00      47   \n",
      "5           True          None                            NaN     479   \n",
      "6          False             1                          96.67      60   \n",
      "7           True          None                            NaN     330   \n",
      "8          False             1                          97.44      39   \n",
      "9           True          None                            NaN     473   \n",
      "10         False             1                          98.33      60   \n",
      "11          True          None                            NaN     281   \n",
      "12         False             1                          98.48      66   \n",
      "13          True          None                            NaN    1593   \n",
      "14         False             1                         100.00      38   \n",
      "15          True          None                            NaN     372   \n",
      "16         False             1                          98.55      69   \n",
      "17          True          None                            NaN    1274   \n",
      "18         False             1                          98.68      76   \n",
      "19          True          None                            NaN     208   \n",
      "20         False             1                         100.00      53   \n",
      "21          True          None                            NaN     554   \n",
      "22         False             1                         100.00      38   \n",
      "23          True          None                            NaN     139   \n",
      "24         False             1                          98.44      64   \n",
      "25          True          None                            NaN     314   \n",
      "26         False             1                          99.01     101   \n",
      "27          True          None                            NaN     650   \n",
      "28         False             1                          98.41     126   \n",
      "29          True          None                            NaN     684   \n",
      "30         False             1                          98.39      62   \n",
      "31          True          None                            NaN     888   \n",
      "32         False             1                          99.42     173   \n",
      "33          True          None                            NaN     379   \n",
      "34         False             1                         100.00      67   \n",
      "35          True          None                            NaN      97   \n",
      "36         False             1                          98.28      58   \n",
      "37          True          None                            NaN    1274   \n",
      "38         False             1                          99.23     130   \n",
      "39          True          None                            NaN    1209   \n",
      "40         False             1                         100.00      46   \n",
      "41          True          None                            NaN     661   \n",
      "42         False             1                          99.00     100   \n",
      "43          True          None                            NaN     562   \n",
      "44         False             1                          96.94      98   \n",
      "45          True          None                            NaN     653   \n",
      "46         False             1                          98.77      81   \n",
      "47          True          None                            NaN    1105   \n",
      "48         False             1                         100.00      47   \n",
      "49          True          None                            NaN     281   \n",
      "50         False             1                         100.00      69   \n",
      "51          True          None                            NaN     480   \n",
      "52         False             1                          98.53      68   \n",
      "53          True          None                            NaN     199   \n",
      "54         False             1                         100.00      92   \n",
      "55          True          None                            NaN     395   \n",
      "56         False             1                          99.36     314   \n",
      "\n",
      "   first_10_nt  last_10_nt                                               path  \\\n",
      "0   gtggtgccca  tacagcaact  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "1   ggtgtgcacg  ctctgtacag  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "2   caagacatct  atgacttata  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "3   ccaggtgtac  gggctcctgc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "4   caggccaagg  acaataccat  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "5   caggtgcgag  ccctcgtggc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "6   cagctacatc  cttccgagtc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "7   agtgtgcagt  atttgcaggc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "8   tagttttgag  tgaatgaaga  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "9   caggtgcgct  cgttggtttc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "10  caggagagca  accgacgata  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "11  gtgtgtcctt  ttttggcagc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "12  cagttgattg  ggattccgtg  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "13  ggtgtgcttt  ttgccccagg  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "14  caagctggtg  accgcggcca  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "15  caggtgtgga  cattattgtg  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "16  cagcggtgaa  atataccaag  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "17  aggtgtgtgt  ccccgtcccc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "18  -aggtcaagg  aacttcgggc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "19  ggtgtgcccc  gctgtttttc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "20  cagccacgcg  tacaggagct  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "21  ggtattcgaa  cccacagccc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "22  gaggaggtgc  ggtcattgaa  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "23  aggtgtggag  ttccgcgccg  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "24  -agtcatcgc  acgtcacggg  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "25  ggtgtggggc  accgtaggtg  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "26  tatctccccg  agaagtcata  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "27  tgcaggtcgc  cttgcgcctg  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "28  cagagtacac  gacataaaga  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "29  aggtgtggct  cctgcttgct  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "30  -aggtgctgt  tcaatattcg  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "31  aggtgtagca  gcgtcgggcg  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "32  -aggggacat  ggccctcctc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "33  ggtgggcccc  tgccgacgca  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "34  gaggccggag  aagatggaag  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "35  aggtgtgccc  ctgtgtcggg  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "36  -agacggacc  tccatcgcga  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "37  aggtgtggcg  cactgttgct  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "38  -aggcgggta  ttgctctccc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "39  ggcatgtccc  ctgcaggttc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "40  gacatcctgt  gacagtgacc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "41  ggtgcgcttg  ccttccacag  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "42  gaccctggcg  caacctactg  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "43  aggtttgtgt  aacccatccc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "44  -agagccatg  cgggcggcca  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "45  aggtgcgcat  cccacgccca  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "46  -aggacgacc  ccaggcccat  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "47  gtgtgttggg  cccttccagg  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "48  cacggctccg  cgtcgaagag  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "49  gcaggtcccg  attcgcttcg  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "50  cagagcggct  caggactccc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "51  ggtccgtccc  tcgcggcggg  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "52  gatttctccg  acagcgtcaa  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "53  gtgtgcgtcg  tcccaggctg  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "54  gacgcggtgc  acgtctgggc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "55  ggtgtgggtc  aaaggtggcc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "56  cacgactact  aagcagagat  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "\n",
      "                                             sequence  \n",
      "0   gtggtgcccatcggcaaggaccaggatggcgtagccggggatggcc...  \n",
      "1   ggtgtgcacgtgtgtgtgggcgcagatgtgctcgcttttttgattg...  \n",
      "2   caagacatctctgaccggcggacgaagatattcgacctgtccctgg...  \n",
      "3   ccaggtgtacccccggctttatggcttatggctcacattcgggctc...  \n",
      "4     caggccaaggagaacgaattggcagacaagatccggaacaataccat  \n",
      "5   caggtgcgagccaccccagtgcagatgtggctttttcgaggaccaa...  \n",
      "6   cagctacatcgagcagctctccagagtcgtcgatgacaacctgcca...  \n",
      "7   agtgtgcagtgtaccttctgtgtgttttgtgtgatgtgtttctggg...  \n",
      "8             tagttttgagcgcgacactttagacgtgctgaatgaaga  \n",
      "9   caggtgcgctgctcattttttcatcgagtctcaggttcatcgtcca...  \n",
      "10  caggagagcaaatggcgatgatgcaggaaaaggcttgccagcggca...  \n",
      "11  gtgtgtcctttcatgctgtccatcaatgctgaattcatgctgtgaa...  \n",
      "12  cagttgattgtcacgtatgatgggcagaagcgactcccactccgcc...  \n",
      "13  ggtgtgctttgtcgtcacccgcagcaagtcaaactgccattggttg...  \n",
      "14             caagctggtgtgcatccaggggattgtcaccgcggcca  \n",
      "15  caggtgtggatgctctgttccttgaacctttggtcgcttggcccat...  \n",
      "16  cagcggtgaagccaaagattcaatttgctgcatacatatgtgagca...  \n",
      "17  aggtgtgtgtgtgtgtgtgtgtgtgtgtgtgtgtgtgtgtgtggtg...  \n",
      "18  -aggtcaagggcgacgcctacacccccgagctgtcttgcagctccg...  \n",
      "19  ggtgtgccccccgcctcttgccttgtgtgttgggcacgtggacaga...  \n",
      "20  cagccacgcggcagcaagttcaccaagttccaggagctgcgtgtac...  \n",
      "21  ggtattcgaagttggatttctggccgagagtgtttctgatttcttt...  \n",
      "22             gaggaggtgccgaaggggtccataccgcggtcattgaa  \n",
      "23  aggtgtggagccccgcggcggccgaactggacccgttgggccgtgt...  \n",
      "24  -agtcatcgcggaggagggcgcgtcctcgcagggcacgccagggga...  \n",
      "25  ggtgtggggctcgcccccccccctcctctctcagttcccttctctg...  \n",
      "26  tatctccccgagtcggcccaggggttcaaggcactgatggccagca...  \n",
      "27  tgcaggtcgccatcacctcttttttggtggcgtgtttgtgtgtgtg...  \n",
      "28  cagagtacacggcgtcccggaagctgcaggaggaggtggaggcgct...  \n",
      "29  aggtgtggctgaggagtgggtgtgatgacgtgcagcggggccacag...  \n",
      "30  -aggtgctgttgttgcagcttgttggcgctggagcgaagtcattgt...  \n",
      "31  aggtgtagcatccggcccggtgtaggtcaccagttggctgtagatg...  \n",
      "32  -aggggacatcaacgtctgcctgatgggggaccccggcgtcgccaa...  \n",
      "33  ggtgggcccctgccccccttgacggtcgcagccgccttcgccgatg...  \n",
      "34  gaggccggagcactggttctggcagacaatggcatttgctgcatcg...  \n",
      "35  aggtgtgccctggcagctgtcctgtgttggcggtactgttggcaca...  \n",
      "36  -agacggaccggacctccatccacgaggtgatggagcagcagacgg...  \n",
      "37  aggtgtggcgctccctctgtcgtctgtccttggcgccactggcaaa...  \n",
      "38  -aggcgggtatcgttacgtctctaaatgcccgggtctcggttctag...  \n",
      "39  ggcatgtccctatcatggcccatcccaggaggaattggtttttgag...  \n",
      "40     gacatcctgtggctgctcctggatgagcctgacgcggacagtgacc  \n",
      "41  ggtgcgcttggcctcctcttttggtgttttttttggaggtgcgccc...  \n",
      "42  gaccctggcgaactttgttgccaagaatcacctcatcacacaggac...  \n",
      "43  aggtttgtgttggcgcctgtgaccgtcgtggggtgcccccacccac...  \n",
      "44  -agagccatgatcgggcaggcgaaggagtattgcccagcggtgccg...  \n",
      "45  aggtgcgcatgttgtccgctgtttgtcccaaggatggtacggcgct...  \n",
      "46  -aggacgaccagatcgccgacgccaccacgtgggtcaccccccgga...  \n",
      "47  gtgtgttgggcggtgttggagttggctttttgttgcgttgcggcag...  \n",
      "48    cacggctccggcggtccaattccatcaacgaggaggacgtcgaagag  \n",
      "49  gcaggtcccgccccgccgcaactcctcccatcgctgccatccttct...  \n",
      "50  cagagcggctgatgaagcagtccaagcgcagcctgattgacgccca...  \n",
      "51  ggtccgtccccgcctcgtcccaccacccttgcccccccccccagcc...  \n",
      "52  gatttctccgcccggatcttcgcgatcatcctggagcagcgggaca...  \n",
      "53  gtgtgcgtcgggttcggggcttgccacacgtcccctctctcccatt...  \n",
      "54  gacgcggtgcgacagcagctcaaggtgcgggggttccgggaggagc...  \n",
      "55  ggtgtgggtccttccccttccccctatccttttttgggtgtaccgg...  \n",
      "56  cacgactactccgaggtgcggttctacacctaaccgcccggggcgg...   \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                       ID  exon_or_intron_number  is_it_intron  \\\n",
      "0  EL_OG0030307_t1_g1_EL_PAM1_GP153_OV05                      1         False   \n",
      "1  EL_OG0030307_t1_g1_EL_PAM1_GP153_OV05                      2          True   \n",
      "2  EL_OG0030307_t1_g1_EL_PAM1_GP153_OV05                      3         False   \n",
      "3  EL_OG0030307_t1_g1_EL_PAM1_GP153_OV05                      4          True   \n",
      "4  EL_OG0030307_t1_g1_EL_PAM1_GP153_OV05                      5         False   \n",
      "\n",
      "  class_of_exon  percent_of_homology_for_exons  length first_10_nt  \\\n",
      "0             1                         100.00     136  gattgaactg   \n",
      "1          None                            NaN    2440  ggccggacca   \n",
      "2             1                          98.55      69  -atgacacca   \n",
      "3          None                            NaN     315  cccacagcca   \n",
      "4             1                         100.00     252  cacagctgtt   \n",
      "\n",
      "   last_10_nt                                               path  \\\n",
      "0  tcgctgcttt  ./alingment_gen_transkrypt/EL_OG0030307_t1_g1_...   \n",
      "1  ggtccgtggc  ./alingment_gen_transkrypt/EL_OG0030307_t1_g1_...   \n",
      "2  ggcccactgt  ./alingment_gen_transkrypt/EL_OG0030307_t1_g1_...   \n",
      "3  ccgccctgcc  ./alingment_gen_transkrypt/EL_OG0030307_t1_g1_...   \n",
      "4  cctgcccccc  ./alingment_gen_transkrypt/EL_OG0030307_t1_g1_...   \n",
      "\n",
      "                                            sequence  \n",
      "0  gattgaactgctgacgggacagcccgcattcccggcctcggaggac...  \n",
      "1  ggccggaccaccttctggtgctcccacaactgtctttcaggatgga...  \n",
      "2  -atgacaccagatgccctcgccctcgccctttgctgcctgcgcaaa...  \n",
      "3  cccacagccagggaggtcctcagggacaggattcttttccccctct...  \n",
      "4  cacagctgttggctgaccccttcttctcggacccggcggtgaagag...   \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                       ID  exon_or_intron_number  is_it_intron  \\\n",
      "0  EH_OG0030308_t2_g2_EH_PAM1_GP153_OV05                      1         False   \n",
      "1  EH_OG0030308_t2_g2_EH_PAM1_GP153_OV05                      2          True   \n",
      "2  EH_OG0030308_t2_g2_EH_PAM1_GP153_OV05                      3         False   \n",
      "3  EH_OG0030308_t2_g2_EH_PAM1_GP153_OV05                      4          True   \n",
      "4  EH_OG0030308_t2_g2_EH_PAM1_GP153_OV05                      5         False   \n",
      "\n",
      "  class_of_exon  percent_of_homology_for_exons  length first_10_nt  \\\n",
      "0             1                          99.32     147  cgaacttcat   \n",
      "1          None                            NaN      65  ggtgtgcctt   \n",
      "2             1                         100.00      57  caagagatct   \n",
      "3          None                            NaN      52  caggtagctc   \n",
      "4             1                         100.00      48  caggcaaaag   \n",
      "\n",
      "   last_10_nt                                               path  \\\n",
      "0  tgaggcaatt  ./alingment_gen_transkrypt/EH_OG0030308_t2_g2_...   \n",
      "1  ttgcatccag  ./alingment_gen_transkrypt/EH_OG0030308_t2_g2_...   \n",
      "2  tgatctgtac  ./alingment_gen_transkrypt/EH_OG0030308_t2_g2_...   \n",
      "3  tctctttggt  ./alingment_gen_transkrypt/EH_OG0030308_t2_g2_...   \n",
      "4  aaacaccatt  ./alingment_gen_transkrypt/EH_OG0030308_t2_g2_...   \n",
      "\n",
      "                                            sequence  \n",
      "0  cgaacttcatgcgaagaagcataatgtgctaacaaggatggccgaa...  \n",
      "1  ggtgtgccttttgcattgtgccttttgcattggacgtgtcgctgcc...  \n",
      "2  caagagatctccgacaggaagacaagaatatttgaattatcgttgg...  \n",
      "3  caggtagctccacgcttgctttataatgtggtccaatggcattctc...  \n",
      "4   caggcaaaagaagatgaattggccgacaagattcggacaaacaccatt   \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                        ID  exon_or_intron_number  \\\n",
      "0   EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                      1   \n",
      "1   EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                      2   \n",
      "2   EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                      3   \n",
      "3   EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                      4   \n",
      "4   EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                      5   \n",
      "5   EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                      6   \n",
      "6   EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                      7   \n",
      "7   EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                      8   \n",
      "8   EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                      9   \n",
      "9   EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                     10   \n",
      "10  EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                     11   \n",
      "11  EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                     12   \n",
      "12  EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                     13   \n",
      "13  EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                     14   \n",
      "14  EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                     15   \n",
      "\n",
      "    is_it_intron class_of_exon  percent_of_homology_for_exons  length  \\\n",
      "0          False             1                          98.86      88   \n",
      "1           True          None                            NaN     142   \n",
      "2          False             1                         100.00      47   \n",
      "3           True          None                            NaN    1298   \n",
      "4          False             1                          98.11      53   \n",
      "5           True          None                            NaN    2446   \n",
      "6          False             1                          97.83      46   \n",
      "7           True          None                            NaN     640   \n",
      "8          False             1                          99.21     127   \n",
      "9           True          None                            NaN     349   \n",
      "10         False             1                          99.34     152   \n",
      "11          True          None                            NaN     357   \n",
      "12         False             1                         100.00      83   \n",
      "13          True          None                            NaN    1670   \n",
      "14         False             1                         100.00      66   \n",
      "\n",
      "   first_10_nt  last_10_nt                                               path  \\\n",
      "0   tggcagcggt  ccatcagcca  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "1   cgccgtgcaa  cgccgtgttc  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "2   caccgcttcg  tgctgcaggg  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "3   ggccaggccg  gcctgtggca  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "4   -agccctctt  ctccgcgcac  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "5   ctccctgtaa  cttactccct  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "6   gaccgcctgc  ttccagcagg  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "7   gtgcgggagg  caacctgcga  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "8   gagatgttca  aagcgggtcc  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "9   ctggtcgagg  tctcgacgct  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "10  gagatcgtct  tgatcggcca  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "11  gcgcccagtg  tcactgcggg  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "12  caattgcggc  tggattctga  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "13  gcgccgaagt  cgcttgcacg  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "14  catctgatga  tgtatgccct  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "\n",
      "                                             sequence  \n",
      "0   tggcagcggtgcgggagcgggcggagtacctcgacgtgatggagga...  \n",
      "1   cgccgtgcaagcacctacataatgactcggcgtgactgcgcgtgac...  \n",
      "2     caccgcttcgccttgcaccagacccagaaggcggcgctgctgcaggg  \n",
      "3   ggccaggccgtggcgcaggccttcgcgctggccctcggcagcccct...  \n",
      "4   -agccctcttcctgaaggcggagctgcccatccgcctcgcccactc...  \n",
      "5   ctccctgtaaggttggtaagactagtcgactagtcgattattcatt...  \n",
      "6      gaccgcctgccccgcggcctcagcggcctcgccgcgttccagcagg  \n",
      "7   gtgcgggaggccaggctgcccttgccggcaaccaacgaccctatcc...  \n",
      "8   gagatgttcacggactccttccgggagctgcgcgcgttccagttct...  \n",
      "9   ctggtcgaggcctgacacaaaatcgacacaccctctccaaaggggt...  \n",
      "10  gagatcgtcttccagggctgccagcagctccgcacgcagctccagc...  \n",
      "11  gcgcccagtgccttgcggggcccactggcagtgctgcccaggaacc...  \n",
      "12  caattgcggctccatgcgcagtacgtgcggggggagagccacgcgg...  \n",
      "13  gcgccgaagtgccatttgttggaacctgaggattttccttaggggt...  \n",
      "14  catctgatgaaggcgtccatagcaaccgccaaacggatggccctcg...   \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "directory = './alingment_gen_transkrypt'\n",
    "\n",
    "for filename in os.listdir(directory):\n",
    "    f = os.path.join(directory, filename)\n",
    "    if not os.path.isfile(f):\n",
    "        continue\n",
    "    if f.endswith(\".aln\"):\n",
    "        #print(\"sciezka: \\t\", f)\n",
    "        cutting_scrap_right(f, 2)\n",
    "        print(\"\\n\\n\\n\\n\")\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae86ba7e-ac79-44f8-a430-738225aa577f",
   "metadata": {},
   "source": [
    "Pierwsza decyzja w przypadku poprawiania kodu na jego prawy odpowiednik (reversed):\n",
    "- nazwy zmiennych które jawnie odnoszą się do pozycji, np. lewe coś tam, pierwszy nukleotyd zostały zmienione na ich właściwe odpowiednki\n",
    "- Pozostałe elementy zostaja spojne - alignment to wciaz zmienna alignment pojawiajaca sie w obu fragmentach kodu.\n",
    "\n",
    "Druga decyzja:\n",
    "- kod bedzie mial strukture:\n",
    "      - przycinanie lewej strony\n",
    "      - przycinanie prawej strony\n",
    "      - wysylka do excela\n",
    "  - i żaden z tych fragmentow nie "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffc6c601-63fc-4f1c-9e5f-a148bc88acbf",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
