{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f037dd59-8d78-4bbc-b1f8-f878afe1f342",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/norbert/miniconda3/lib/python3.11/site-packages/Bio/pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from Bio import AlignIO, SeqIO, Seq, pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e7f5ae9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "min_length_aligned_sequence = 30 #Minimal lenght of sequence which could be an exon\n",
    "extreme_homology = 0.97 #percentage of homology of sequence, treshold #I assume two faulty aligned nucleotides per 100 (98%) and one more nt because sometimes latest nt can move from end of one sequence to beginning next sequence\n",
    "middle_homology = 0.9 #used for classification of internal exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "038e46c9-9a33-498a-a7fc-618d1fa05a39",
   "metadata": {},
   "outputs": [],
   "source": [
    "#uporzadkowac kod tworzac w nim funkcje []\n",
    "#potrzebne indexy rozpoczecia intronow i eksonow liczac od poczatku genu bedacego inputem czyli to +-500 do gff\n",
    "#+ jeszcze numeracja intron i ekson osobno - ale w jednej kolumnnie - E1 I1 E2 I2 [CHECK]\n",
    "#obrobic jakos ladnie ten kod - definiowac funkcje ktore sie powtarzaja\n",
    "#cel na chwile potem:\n",
    "    #wziac pliki wyslane przez pawla i w bio pythonie je polaczyc. Jeden plik fasta zawiera po trzy scaffoldy dla kazdego gatunku oraz trzy odczyty trinity. Biopython musi je wziac ze soba i polaczyc wywalajac gapy.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "e6d4c5ee-799f-427e-839a-f95917b6a430",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " \n",
      " EL_OG0030308_t2_g2_EL_PAM1_GP153_OV05 sequence is too short, sequence has not exons or other fault. \n",
      "\n",
      "                                        ID exon_or_intron_number is_it_intron  \\\n",
      "1    EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                    E1          0.0   \n",
      "2    EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                    I1          1.0   \n",
      "3    EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                    E2          0.0   \n",
      "4    EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                    I2          1.0   \n",
      "5    EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                    E3          0.0   \n",
      "..                                     ...                   ...          ...   \n",
      "144  EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                    E6          0.0   \n",
      "145  EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                    I6          1.0   \n",
      "146  EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                    E7          0.0   \n",
      "147  EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                    I7          1.0   \n",
      "148  EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                    E8          0.0   \n",
      "\n",
      "    class_of_exon  percent_of_homology_for_exons  length first_10_nt  \\\n",
      "1               1                          97.92    96.0  ctacacgtac   \n",
      "2            None                            NaN   102.0  gcccgccgcc   \n",
      "3               1                         100.00   145.0  gatcgagctg   \n",
      "4            None                            NaN   238.0  gaccgggcgg   \n",
      "5               1                         100.00    72.0  atgtcgccgg   \n",
      "..            ...                            ...     ...         ...   \n",
      "144             1                          99.35   154.0  ctgagatcgt   \n",
      "145          None                            NaN   356.0  gcgcccagtg   \n",
      "146             1                         100.00    84.0  gcaattgcgg   \n",
      "147          None                            NaN  1669.0  gcgccgaagt   \n",
      "148             1                         100.00    67.0  gcatctgatg   \n",
      "\n",
      "     last_10_nt                                               path  \\\n",
      "1    ggctgctcct  /home/norbert/mrdn/euglena/kod_i_pliki/alingme...   \n",
      "2    cggcgggggc  /home/norbert/mrdn/euglena/kod_i_pliki/alingme...   \n",
      "3    ccgctgcttt  /home/norbert/mrdn/euglena/kod_i_pliki/alingme...   \n",
      "4    gcccgcggtc  /home/norbert/mrdn/euglena/kod_i_pliki/alingme...   \n",
      "5    caccgtggca  /home/norbert/mrdn/euglena/kod_i_pliki/alingme...   \n",
      "..          ...                                                ...   \n",
      "144  tgatcggcca  /home/norbert/mrdn/euglena/kod_i_pliki/alingme...   \n",
      "145  gtcactgcgg  /home/norbert/mrdn/euglena/kod_i_pliki/alingme...   \n",
      "146  tggattctga  /home/norbert/mrdn/euglena/kod_i_pliki/alingme...   \n",
      "147  gcgcttgcac  /home/norbert/mrdn/euglena/kod_i_pliki/alingme...   \n",
      "148  tgtatgccct  /home/norbert/mrdn/euglena/kod_i_pliki/alingme...   \n",
      "\n",
      "                                              sequence  \n",
      "1    ctacacgtacaagcacgtgaacgtcatggcgccggaggtccaaagc...  \n",
      "2    gcccgccgcccggggttcggtgcccccggctgtcaggccgttgggt...  \n",
      "3    gatcgagctgctgaccggccagccggcgttccccgtctcggaggac...  \n",
      "4    gaccgggcggcctcaccatggcagtgccacatcctgggccaacaga...  \n",
      "5    atgtcgccggacgccctctcgctggccctccgctgcctcgccaagt...  \n",
      "..                                                 ...  \n",
      "144  ctgagatcgtcttccagggctgccagcagctccgcacgcagctcca...  \n",
      "145  gcgcccagtgccttgcggggcccactggcagtgctgcccaggaacc...  \n",
      "146  gcaattgcggctccatgcgcagtacgtgcggggggagagccacgcg...  \n",
      "147  gcgccgaagtgccatttgttggaacctgaggattttccttaggggt...  \n",
      "148  gcatctgatgaaggcgtccatagcaaccgccaaacggatggccctc...  \n",
      "\n",
      "[148 rows x 10 columns]\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "File already exists. Do you want overwrite? y/n:  k\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Clarify your answer. Nothing has done.\n"
     ]
    }
   ],
   "source": [
    "def cutting_scrap_right(directory, acceptable_gap_length):\n",
    "    data_frames = []\n",
    "    gaps_signs = \"-\" * acceptable_gap_length #maximum length of gaps in sequence in one exon's sequence\n",
    "    #print(\"helo\")\n",
    "    \n",
    "    def extracting_strands_from_alignment(alignment):\n",
    "        seq1 = alignment[0].seq #transcript sequence\n",
    "        seq2 = alignment[1].seq #gene sequence\n",
    "      #  print(\"helo2\")\n",
    "        return seq1, seq2\n",
    "\n",
    "    #def setting_first_nucleotide_pair(alignment): #ta funkcja ucina nam 6 odczytow #to nam nie działa więc nie brnę dalej\n",
    "        #if \"-\" in alignment[:, 0]: #deleting mismatches at the first nucleotide\n",
    "            #return alignment[:, 1:]\n",
    "        #else:\n",
    "            #return alignment\n",
    "        #print(\"helo3\")\n",
    "            #print(f\"alignment after cut:\\n{alignment}\") #chekpoint\n",
    "\n",
    "\n",
    "    #def setting_last_nuckleotide_pair(alignment): #to nam nie działa więc nie brnę dalej\n",
    "        #if \"-\" in alignment[:, -1]: #deleting mismatches at the end\n",
    "            #alignment = alignment[:, :-1]\n",
    "            #return alignment \n",
    "        #else:\n",
    "            #return alignment #print(f\"alignment check:\\n{alignment[:, -50:-1]}\") checkpoint\n",
    "                    \n",
    "\n",
    "    def finding_gap_index(sequence, acceptable_gap_length):\n",
    "        return sequence.find(\"-\" * acceptable_gap_length) \n",
    "\n",
    "\n",
    "    def reversed_finding_gap_index(sequence, acceptable_gap_length):\n",
    "        return sequence.rfind(\"-\" * acceptable_gap_length) + acceptable_gap_length #now we have to add length of gap to index, because python starts counting from 0 nt, and we just change direction of counting.\n",
    "\n",
    "    def setting_distance_between_first_nt_and_gap(is_it_exon):\n",
    "        if is_it_exon == True:\n",
    "            if first_gap_position_seq2 < 0: #when in second strand there are no gaps - it means, second strand is correctly prepared\n",
    "                distance_between_first_nt_and_gap = first_gap_position_seq1\n",
    "            else:\n",
    "                distance_between_first_nt_and_gap = min(first_gap_position_seq1, first_gap_position_seq2)\n",
    "            return distance_between_first_nt_and_gap\n",
    "        elif is_it_exon == False:\n",
    "            if last_gap_position_seq2 < 0: #it means, if there is no gaps in second sequention, gap position is gap position in seq1. Additional, rfind() function return \"-1\" if it won't find any gaps in seq\n",
    "                distance_between_first_nt_and_gap = (len(seq1) - last_gap_position_seq1) -1\n",
    "                #distance_between_first_nt_and_gap = distance_between_first_nt_and_gap - 1 #correction to counting from 0th index\n",
    "                #print(\"pierwszy\", distance_between_first_nt_and_gap) #checkpoint\n",
    "            else:\n",
    "                distance_between_first_nt_and_gap = (len(seq1) - max(last_gap_position_seq1, last_gap_position_seq2))\n",
    "            return distance_between_first_nt_and_gap\n",
    "                #print(\"drugi\", distance_between_first_nt_and_gap) #checkpoint\n",
    "\n",
    "    \n",
    "    def save_to_file(final_data_frame, filename):\n",
    "        if os.path.isfile(filename):\n",
    "            user_input = input(\"File already exists. Do you want overwrite? y/n: \")\n",
    "            if user_input == \"n\":\n",
    "                base_name, ext = os.path.splitext(filename)\n",
    "                i = 1\n",
    "                while os.path.isfile(f\"{base_name}_{i}{ext}\"):\n",
    "                    i += 1\n",
    "                filename = f\"{base_name}_{i}{ext}\"\n",
    "                final_data_frame.to_csv(filename, sep = \"\\t\")\n",
    "                \n",
    "            elif user_input == \"y\":\n",
    "                final_data_frame.to_csv(filename, sep = \"\\t\")\n",
    "            else:\n",
    "                print(\"Clarify your answer. Nothing has done.\")\n",
    "                \n",
    "        else:\n",
    "            final_data_frame.to_csv(filename, sep = \"\\t\")\n",
    "                     \n",
    "    \n",
    "    for filename in os.listdir(directory):\n",
    "        #print(\"helo4\")\n",
    "        file = os.path.join(directory, filename)\n",
    "        if not os.path.isfile(file):\n",
    "            continue\n",
    "        if file.endswith(\".aln\"):\n",
    "            alignment = AlignIO.read(file, \"clustal\")\n",
    "           \n",
    "        \n",
    "            while True:\n",
    "                is_it_exon = True\n",
    "                #alignment = setting_first_nucleotide_pair(alignment)\n",
    "                first_nucleotides_pair = alignment[:, 0]#first nucleotides in both strands as variable\n",
    "                if \"-\" in first_nucleotides_pair: #deleting mismatches at the start \n",
    "                    alignment = alignment[:, 1:]\n",
    "                    #print(f\"alignment after cut:\\n{alignment}\") #chekpoint\n",
    "                    continue\n",
    "\n",
    "                \n",
    "                seq1, seq2 = extracting_strands_from_alignment(alignment)\n",
    "\n",
    "                \n",
    "                first_gap_position_seq1 = finding_gap_index(seq1, acceptable_gap_length)\n",
    "                first_gap_position_seq2 = finding_gap_index(seq2, acceptable_gap_length)\n",
    "                #print(f\"first gap position in first sequence:{first_gap_position_seq1}, in 2nd sequence: {first_gap_position_seq2}\") #checkpoint\n",
    "\n",
    "                \n",
    "                #if first_gap_position_seq2 < 0: #avoiding faults when find() function returns \"-1\" as finding nothing\n",
    "                 #   distance_between_first_nt_and_gap = first_gap_position_seq1\n",
    "          #          print(\"helo6\")\n",
    "                #else:\n",
    "                 #   distance_between_first_nt_and_gap = min(first_gap_position_seq1, first_gap_position_seq2)\n",
    "           #         print(\"helo7\")\n",
    "                distance_between_first_nt_and_gap = setting_distance_between_first_nt_and_gap(is_it_exon)\n",
    "                \n",
    "                local_alignment = pairwise2.align.localxx(seq1[:distance_between_first_nt_and_gap], seq2[:distance_between_first_nt_and_gap], one_alignment_only = True)\n",
    "            #    print(\"helo8\")\n",
    "                #\"xx\"  means no gap penalty while opening gaps or longering them and no penalties for mismatch. Just pure score of alignment to count homology\n",
    "                #print(\"local_alignment:\", local_alignment) #checkpoint\n",
    "               \n",
    "                try: \n",
    "                    local_homology_percentage = (local_alignment[0].score / distance_between_first_nt_and_gap) \n",
    "                except IndexError:\n",
    "                    if len(seq1) < 100:\n",
    "                        print(f\"\\n \\n {alignment[0].id} sequence is too short, sequence has not exons or other fault. \\n\")\n",
    "                        break\n",
    "                    continue\n",
    "                \n",
    "                \n",
    "                if local_homology_percentage <= extreme_homology:\n",
    "                    #print(f\"\\n Too low score of homology. \\n alignment: {alignment[:, :50]} \\n score: {local_alignment[0].score}, distance: {distance_between_first_nt_and_gap}, percent: {round(local_homology_percentage, 2)*100}%, \\n {local_alignment}\")\n",
    "                    alignment = alignment[:, distance_between_first_nt_and_gap:] \n",
    "                    #print(\"helo9\")\n",
    "                    \n",
    "                else:\n",
    "                    if distance_between_first_nt_and_gap <= min_length_aligned_sequence:\n",
    "                        #print(f\"\\n alignment is too short. \\n distance between sequences {distance_between_first_nt_and_gap}, \\n alignment: {alignment}, \\n score: {local_alignment[0].score}, distance: {distance_between_first_nt_and_gap}, procentowo: {round(local_homology_percentage, 2)*100}%, \\n {local_alignment}\")\n",
    "                        alignment = alignment[:, distance_between_first_nt_and_gap:]\n",
    "                   #     print(\"helo10\")\n",
    "                    else:\n",
    "                        #print(f\"Sequence {alignment[0].id} cut properly from LEFT side, \\n distance between sequences: {distance_between_first_nt_and_gap}\"), alignment: {alignment[:, :50]},\n",
    "               #         print(\"helo11\")\n",
    "                        break\n",
    "\n",
    "               \n",
    "        \n",
    "            \n",
    "        ######################### TERAZ OD PRAWEJ DO LEWEJ #######################  \n",
    "        \n",
    "            \n",
    "            while True:\n",
    "                is_it_exon = False\n",
    "                #last_nucleotides_pair = setting_last_nuckleotide_pair(alignment)\n",
    "                last_nucleotides_pair = alignment[:, -1]#last nucleotides in both strands as variable\n",
    "                if \"-\" in last_nucleotides_pair: #deleting mismatches at the end\n",
    "                    alignment = alignment[:, :-1]\n",
    "                    #print(f\"alignment check:\\n{alignment[:, -50:-1]}\") checkpoint\n",
    "                    continue\n",
    "                        \n",
    "                            \n",
    "                seq1, seq2 = extracting_strands_from_alignment(alignment)\n",
    "\n",
    "                last_gap_position_seq1 = reversed_finding_gap_index(seq1, acceptable_gap_length)\n",
    "                last_gap_position_seq2 = reversed_finding_gap_index(seq2, acceptable_gap_length)\n",
    "                #last_gap_position_seq1 = seq1.rfind(gaps_signs)  #Counting distance: #getting index of first gap. \n",
    "                #last_gap_position_seq1 = last_gap_position_seq1 + acceptable_gap_length\n",
    "                \n",
    "                #last_gap_position_seq2 = seq2.rfind(gaps_/igns)  \n",
    "                #last_gap_position_seq2 = last_gap_position_seq2 + acceptable_gap_length #now we have to add length of gap to index, because python starts counting from 0 nt, and we just change direction of counting.\n",
    "                #print(alignment[0].id)\n",
    "                #print(\"seq1:\", last_gap_position_seq1, \"seq2:\", last_gap_position_seq2) #checkpoint\n",
    "                \n",
    "                \n",
    "                #if last_gap_position_seq2 < 0: #it means, if there is no gaps in second sequention, gap position is gap position in seq1. Additional, rfind() function return \"-1\" if it won't find any gaps in seq\n",
    "                 #   distance_between_first_nt_and_gap = (len(seq1) - last_gap_position_seq1)\n",
    "                  #  distance_between_first_nt_and_gap = distance_between_first_nt_and_gap - 1 #correction to counting from 0th index\n",
    "                   # #print(\"pierwszy\", distance_between_first_nt_and_gap)\n",
    "                #else:\n",
    "                 #   distance_between_first_nt_and_gap = (len(seq1) - max(last_gap_position_seq1, last_gap_position_seq2))\n",
    "                    #print(\"drugi\", distance_between_first_nt_and_gap)\n",
    "                distance_between_first_nt_and_gap = setting_distance_between_first_nt_and_gap(is_it_exon)\n",
    "                \n",
    "                local_alignment = pairwise2.align.localxx(seq1[-distance_between_first_nt_and_gap:], seq2[-distance_between_first_nt_and_gap:], one_alignment_only = True)\n",
    "                    #\"xx\" means no gap penalty while opening gaps or longering them and no penalties for mismatch. Just pure score of alignment\n",
    "                #print(\"local_alignment:\", local_alignment) #checkpoint\n",
    "                \n",
    "                \n",
    "        \n",
    "                try:\n",
    "                    if distance_between_first_nt_and_gap != 0:\n",
    "                        local_homology_percentage = (local_alignment[0].score / distance_between_first_nt_and_gap) \n",
    "                    else:\n",
    "                        break\n",
    "                except IndexError:\n",
    "                    if len(seq1) < 100:\n",
    "                        print(f\"\\n \\n {alignment[0].id} sequence is too short or sequence has not exons. \\n\")\n",
    "                        break\n",
    "                    continue\n",
    "        \n",
    "                \n",
    "                if local_homology_percentage <= extreme_homology:\n",
    "                    #print(f\"\\n Too low homology! \\n alignment: {alignment[:, -50:]} \\n score: {local_alignment[0].score}, \\distance: {distance_between_first_nt_and_gap}, %%%: {round(local_homology_percentage, 2)*100}%, \\n {local_alignment}\")\n",
    "                    alignment = alignment[:, :-distance_between_first_nt_and_gap] \n",
    "                    \n",
    "                else:\n",
    "                    if distance_between_first_nt_and_gap <= min_length_aligned_sequence:\n",
    "                        #print(f\"\\n Alignment is too short!. \\n distance: {distance_between_first_nt_and_gap}, \\n alignment: {alignment}, \\n score: {local_alignment[0].score}, distance: {distance_between_first_nt_and_gap}, %%%: {round(local_homology_percentage, 2)*100}%, \\n {local_alignment}\")\n",
    "                        alignment = alignment[:, :-distance_between_first_nt_and_gap]\n",
    "                    else:\n",
    "                        #print(f\"Sequence: {alignment[0].id} cut properly from RIGHT side\")\n",
    "                        #print(f\"\"\"\\n \\n \\n victory prawo! \\n distance between sequences {distance_between_first_nt_and_gap},\n",
    "                        #alignment: {alignment[:, -50:]}, \n",
    "                        #score: {local_alignment[0].score}, distance: {distance_between_first_nt_and_gap}, \n",
    "                        #percent: {round(local_homology_percentage, 4)*100}%, \\n \\n local alignment to confirm: \\n{local_alignment}\"\"\")\n",
    "                        break\n",
    "\n",
    "        \n",
    "        \n",
    "            \n",
    "            nucleotides = [\"a\", \"t\", \"g\", \"c\", \"A\", \"T\", \"G\", \"C\"]\n",
    "            index = 0\n",
    "            id = []\n",
    "            is_it_intron = []\n",
    "            class_of_exon = []\n",
    "            exon_or_intron_number = []\n",
    "            percent_of_homology_for_exons = []\n",
    "            length = []\n",
    "            first_10_nt = []\n",
    "            last_10_nt = []\n",
    "            sequence = []\n",
    "            path = []\n",
    "            \n",
    "            while True:\n",
    "                try: \n",
    "                    if distance_between_first_nt_and_gap != 0:\n",
    "                        local_homology_percentage = (local_alignment[0].score / distance_between_first_nt_and_gap) #that sign \"[]\" becouse local_alignment's type is list\n",
    "                    else:\n",
    "                        break\n",
    "                except IndexError:\n",
    "                    if len(seq1) < 100:\n",
    "                        print(f\"\\n \\n {alignment[0].id} sequence is too short or sequence has not exons. \\n\")\n",
    "                        break\n",
    "                    continue\n",
    "                \n",
    "                if len(seq1) > 1 and seq1[0] != gaps_signs:#mining data from exons\n",
    "                    index += 1\n",
    "                    last_nt_exon_index = seq1.find(gaps_signs) \n",
    "                \n",
    "                    if last_nt_exon_index == -1: #it means if it is last exon, if find() don't find any gap signs at the end\n",
    "                        last_nt_exon_index = len(seq1)\n",
    "                                \n",
    "                    alignment_of_exon = pairwise2.align.localxx(seq1[:last_nt_exon_index], seq2[:last_nt_exon_index], one_alignment_only = True)\n",
    "                        #print( \"\\n alignment since start to the end of exon:\", alignment_of_exon[0])\n",
    "                    percent_of_homology = alignment_of_exon[0].score / last_nt_exon_index\n",
    "\n",
    "                        #print(\"exon analising finished\")\n",
    "                        #print(f\"score alignmentu = , {alignment_of_exon[0].score}\\n seq1 = {seq1}, \\n alignment = , {alignment_of_exon[0]}\")\n",
    "                            \n",
    "            \n",
    "\n",
    "                    id.append(alignment[0].id)\n",
    "                    exon_or_intron_number.append(f\"E{index}\")\n",
    "                    is_it_intron.append(False)\n",
    "                    percent_of_homology_for_exons.append(round(percent_of_homology * 100, 2))\n",
    "                    length.append(last_nt_exon_index)\n",
    "                    first_10_nt.append(str(seq1[:10]))\n",
    "                    last_10_nt.append(str(seq1[last_nt_exon_index-10 : last_nt_exon_index])) \n",
    "                    sequence.append(str(seq1[:last_nt_exon_index]))\n",
    "                    path.append(str(file))\n",
    "                            \n",
    "                    if percent_of_homology >= 0.9 and last_nt_exon_index >= min_length_aligned_sequence :\n",
    "                        class_of_exon.append(\"1\")\n",
    "                    elif percent_of_homology < 0.9 and percent_of_homology > 0 and last_nt_exon_index < min_length_aligned_sequence:\n",
    "                        class_of_exon.append(\"2\")\n",
    "                    else:\n",
    "                        class_of_exon.append(\"3\")\n",
    "            \n",
    "                    seq1 = seq1[last_nt_exon_index:]\n",
    "                    seq2 = seq2[last_nt_exon_index:]\n",
    "                    #print(\"analising exon finished\")\n",
    "            \n",
    "                            \n",
    "            \n",
    "                if len(seq1) > 1 and seq1[0] == \"-\":\n",
    "            \n",
    "                    for i, nukleotide in enumerate(seq1):\n",
    "                        if nukleotide not in nucleotides:\n",
    "                            last_nt_intron_index = i+1\n",
    "                        else:\n",
    "                            break\n",
    "                            \n",
    "                    id.append(alignment[0].id)\n",
    "                    exon_or_intron_number.append(f\"I{index}\")\n",
    "                    is_it_intron.append(True) \n",
    "                    percent_of_homology_for_exons.append(None)\n",
    "                    length.append(last_nt_intron_index)\n",
    "                    first_10_nt.append(str(seq2[:10]))\n",
    "                    last_10_nt.append(str(seq2[last_nt_intron_index-10 : last_nt_intron_index]))\n",
    "                    sequence.append(str(seq2[:last_nt_intron_index]))\n",
    "                    path.append(str(file))\n",
    "                    class_of_exon.append(None)\n",
    "        \n",
    "                    seq1 = seq1[last_nt_intron_index:]\n",
    "                    seq2 = seq2[last_nt_intron_index:]\n",
    "                    \n",
    "                else:\n",
    "                    break\n",
    "                        \n",
    "            dictionary = {\"ID\":id,\n",
    "                          \"exon_or_intron_number\":exon_or_intron_number, \n",
    "                          \"is_it_intron\":is_it_intron, \n",
    "                          \"class_of_exon\":class_of_exon, \n",
    "                          \"percent_of_homology_for_exons\":percent_of_homology_for_exons, \n",
    "                          \"length\":length,\n",
    "                          \"first_10_nt\" : first_10_nt,\n",
    "                          \"last_10_nt\" : last_10_nt, \n",
    "                          \"path\":path, \n",
    "                          \"sequence\" : sequence}\n",
    "            data_frame = pd.DataFrame(dictionary)\n",
    "            #print(data_frame)\n",
    "        data_frames.append(data_frame)\n",
    "        #print(data_frames)\n",
    "\n",
    "\n",
    "    final_data_frame = pd.concat(data_frames, ignore_index=True)\n",
    "    final_data_frame.index = np.arange(1, len(final_data_frame) + 1)\n",
    "    print(final_data_frame)\n",
    "    \n",
    "    save_to_file(final_data_frame, \"przyciete_introny.tsv\")\n",
    "\n",
    "\n",
    "cutting_scrap_right('/home/norbert/mrdn/euglena/kod_i_pliki/alingment_gen_transkrypt/', 2)\n",
    "\n",
    "\n",
    "#zapisac w formacie gff\n",
    " #cele na teraz:\n",
    "\n",
    "#mamy dwie opcje, albo ręcznie ustawiamy to dla gff przez np. tworzenie list, albo szukamy informacji jak obecne dane i ich rozszerzenia mogą posłużyć do konwertowania na formaty bardziej przyjazne gff. Ma to większy próg wejścia ale więcej uczy i minimalizuje błędy\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2a04e46-1593-4d3e-9423-54efbcfe4bbe",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
