{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f037dd59-8d78-4bbc-b1f8-f878afe1f342",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/norbert/miniconda3/lib/python3.11/site-packages/Bio/pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from Bio import AlignIO, SeqIO, Seq, pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e7f5ae9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "min_length_aligned_sequence = 30 #Minimal lenght of sequence which could be an exon\n",
    "extreme_homology = 0.97 #percentage of homology of sequence, treshold #I assume two faulty aligned nucleotides per 100 (98%) and one more nt because sometimes latest nt can move from end of one sequence to beginning next sequence\n",
    "middle_homology = 0.9 #nothing at the moment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d05490af-896a-4600-92e8-008e685894a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "path11 = './alingment_gen_transkrypt//EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05.aln'\n",
    "path12 = './alingment_gen_transkrypt//EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05.aln'\n",
    "path13 = './alingment_gen_transkrypt//EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05.aln'\n",
    "path21 = './alingment_gen_transkrypt//EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05.aln'\n",
    "path22 = './alingment_gen_transkrypt//EH_OG0030308_t2_g2_EH_PAM1_GP153_OV05.aln'\n",
    "path23 = './alingment_gen_transkrypt//EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05.aln'\n",
    "path31 = './alingment_gen_transkrypt//EL_OG0030307_t1_g1_EL_PAM1_GP153_OV05.aln'\n",
    "path32 = './alingment_gen_transkrypt//EL_OG0030308_t2_g2_EL_PAM1_GP153_OV05.aln'\n",
    "path33 = './alingment_gen_transkrypt//EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05.aln'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "479c14c9-478a-4948-986d-16e849c2e4af",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def cutting_scrap_right(file, acceptable_gap_length):\n",
    "    \n",
    "    gaps_signs = \"-\" * acceptable_gap_length #maximum length of gaps one by one in sequence\n",
    "    \n",
    "    alignment = AlignIO.read(file, \"clustal\")\n",
    "    \n",
    "        \n",
    "        \n",
    "    while True:\n",
    "        first_nucleotides_pair = alignment[:, 0]#first nucleotides in both strands as variable\n",
    "        if \"-\" in first_nucleotides_pair: #deleting mismatches at the start #do tej pory wszystko git. program odnajduje znak \"-\" w pierwszej. parze\n",
    "            alignment = alignment[:, 1:]\n",
    "            #print(f\"alignment po obrobce:\\n{alignment}\") #chekpoint\n",
    "            continue\n",
    "    \n",
    "            \n",
    "        \n",
    "        #changing extension\n",
    "        seq1 = alignment[0].seq #transcript sequence\n",
    "        seq2 = alignment[1].seq #gene sequence\n",
    "\n",
    "        \n",
    "        #Counting distance:\n",
    "        first_gaps_position_seq1 = seq1.find(gaps_signs) #first gap inddex. There is also included gap in genomic sequence - made it intentionally\n",
    "        first_gaps_position_seq2 = seq2.find(gaps_signs)  #this code prints alignment since 1st to certain number (without gaps at the end!)\n",
    "        #print(f\"first gap position in first sequence:{first_gaps_position_seq1}, in 2nd sequence: {first_gaps_position_seq2}\") #checkpoint\n",
    "        \n",
    "        if first_gaps_position_seq2 < 0:\n",
    "            distance_between_first_nt_and_gap = first_gaps_position_seq1\n",
    "        else:\n",
    "            distance_between_first_nt_and_gap = min(first_gaps_position_seq1, first_gaps_position_seq2)\n",
    "\n",
    "        \n",
    "        local_alignment = pairwise2.align.localxx(seq1[:distance_between_first_nt_and_gap], seq2[:distance_between_first_nt_and_gap], \\\n",
    "                                                    one_alignment_only = True)#score_only = True print only score\n",
    "        #\"xx\"  means no gap penalty while opening gaps or longering them and no penalties for mismatch. Just pure score of alignment to count homology\n",
    "        #print(\"local_alignment:\", local_alignment) #checkpoint\n",
    "       \n",
    "        try: \n",
    "            local_homology_percentage = (local_alignment[0].score / distance_between_first_nt_and_gap) \n",
    "        except IndexError:\n",
    "            #print(\"index error\", len(seq1))\n",
    "            if len(seq1) < 100:\n",
    "                print(f\"\\n \\n {alignment[0].id} sequence is too short, sequence has not exons or other fault. \\n\")\n",
    "                break\n",
    "            continue\n",
    "        \n",
    "        \n",
    "        if local_homology_percentage <= extreme_homology:\n",
    "            #print(f\"\\n Too low score of homology. \\n alignment: {alignment[:, :50]} \\n score: {local_alignment[0].score}, distance: {distance_between_first_nt_and_gap}, percent: {round(local_homology_percentage, 2)*100}%, \\n {local_alignment}\")\n",
    "            alignment = alignment[:, distance_between_first_nt_and_gap:] \n",
    "            \n",
    "        else:\n",
    "            if distance_between_first_nt_and_gap <= min_length_aligned_sequence:\n",
    "                #print(f\"\\n alignment is too short. \\n distance between sequences {distance_between_first_nt_and_gap}, \\n alignment: {alignment}, \\n score: {local_alignment[0].score}, distance: {distance_between_first_nt_and_gap}, procentowo: {round(local_homology_percentage, 2)*100}%, \\n {local_alignment}\")\n",
    "                alignment = alignment[:, distance_between_first_nt_and_gap:]\n",
    "            else:\n",
    "                #print(f\"Sequence {alignment[0].id} cut properly from LEFT side\")\n",
    "                #print(f\"\"\"\\n \\n \\n victory left side! \\n distance between sequences: {distance_between_first_nt_and_gap},\n",
    "                #alignment: {alignment[:, :50]}, \n",
    "                #score: {local_alignment[0].score}, dystans: {distance_between_first_nt_and_gap}, \n",
    "                #percent: {round(local_homology_percentage, 4)*100}%, \\n \\n local alignment to confirm: \\n{local_alignment}\"\"\")\n",
    "                break\n",
    "        \n",
    "\n",
    "    \n",
    "######################### TERAZ OD PRAWEJ DO LEWEJ #######################  \n",
    "\n",
    "    \n",
    "    while True:\n",
    "        last_nucleotides_pair = alignment[:, -1]#last nucleotides in both strands as variable\n",
    "        if \"-\" in last_nucleotides_pair: #deleting mismatches at the end\n",
    "            alignment = alignment[:, :-1]\n",
    "            #print(f\"alignment check:\\n{alignment[:, -50:-1]}\") checkpoint\n",
    "            continue\n",
    "                \n",
    "                    \n",
    "        #changing extension\n",
    "        seq1 = alignment[0].seq #transcript sequence\n",
    "        seq2 = alignment[1].seq #gene sequence\n",
    "        \n",
    "       \n",
    "        last_gaps_position_seq1 = seq1.rfind(gaps_signs)  #Counting distance: #getting index of first gap. There is also included gap in genomic sequence - made it intentionally\n",
    "        last_gaps_position_seq1 = last_gaps_position_seq1 + acceptable_gap_length\n",
    "        \n",
    "        last_gaps_position_seq2 = seq2.rfind(gaps_signs)  #this code prints alignment since 1st to certain number (without gaps at the end!)\n",
    "        last_gaps_position_seq2 = last_gaps_position_seq2 + acceptable_gap_length #now we have to add length of gap to index, because python starts counting from 0 nt, and we just change direction of counting.\n",
    "        #print(\"seq1:\", last_gaps_position_seq1, \"seq2:\", last_gaps_position_seq2) checkpoint\n",
    "        \n",
    "        \n",
    "        if last_gaps_position_seq2 < 0: #it means, if there is no gaps in second sequention, gap position is gap position in seq1. Additional, rfind() function return \"-1\" if it won't find any gaps in seq\n",
    "            distance_between_first_nt_and_gap = (len(seq1) - last_gaps_position_seq1)\n",
    "            distance_between_first_nt_and_gap = distance_between_first_nt_and_gap - 1 #correction to counting from 0th index\n",
    "            #print(\"pierwszy\", distance_between_first_nt_and_gap)\n",
    "        else:\n",
    "            distance_between_first_nt_and_gap = (len(seq1) - max(last_gaps_position_seq1, last_gaps_position_seq2))\n",
    "            #print(\"drugi\", distance_between_first_nt_and_gap)\n",
    "            \n",
    "            \n",
    "            local_alignment = pairwise2.align.localxx(seq1[-distance_between_first_nt_and_gap:], seq2[-distance_between_first_nt_and_gap:], \\\n",
    "                                                    one_alignment_only = True) #score_only = True print only score \n",
    "            #\"xx\" means no gap penalty while opening gaps or longering them and no penalties for mismatch. Just pure score of alignment\n",
    "        #print(\"local_alignment:\", local_alignment) #checkpoint\n",
    "        \n",
    "        \n",
    "\n",
    "        try:\n",
    "            local_homology_percentage = (local_alignment[0].score / distance_between_first_nt_and_gap) \n",
    "        except IndexError:\n",
    "            if len(seq1) < 100:\n",
    "                print(f\"\\n \\n {alignment[0].id} sequence is too short or sequence has not exons. \\n\")\n",
    "                break\n",
    "            continue\n",
    "\n",
    "        \n",
    "        if local_homology_percentage < extreme_homology:\n",
    "            #print(f\"\\n Too low homology! \\n alignment: {alignment[:, -50:]} \\n score: {local_alignment[0].score}, \\distance: {distance_between_first_nt_and_gap}, %%%: {round(local_homology_percentage, 2)*100}%, \\n {local_alignment}\")\n",
    "            alignment = alignment[:, :-distance_between_first_nt_and_gap] \n",
    "            \n",
    "        else:\n",
    "            if distance_between_first_nt_and_gap < min_length_aligned_sequence:\n",
    "                #print(f\"\\n Alignment is too short!. \\n distance: {distance_between_first_nt_and_gap}, \\n alignment: {alignment}, \\n score: {local_alignment[0].score}, distance: {distance_between_first_nt_and_gap}, %%%: {round(local_homology_percentage, 2)*100}%, \\n {local_alignment}\")\n",
    "                alignment = alignment[:, :-distance_between_first_nt_and_gap]\n",
    "            else:\n",
    "                #print(f\"Sequence: {alignment[0].id} cut properly from RIGHT side\")\n",
    "                #print(f\"\"\"\\n \\n \\n victory prawo! \\n distance between sequences {distance_between_first_nt_and_gap},\n",
    "                #alignment: {alignment[:, -50:]}, \n",
    "                #score: {local_alignment[0].score}, distance: {distance_between_first_nt_and_gap}, \n",
    "                #percent: {round(local_homology_percentage, 4)*100}%, \\n \\n local alignment to confirm: \\n{local_alignment}\"\"\")\n",
    "                break\n",
    "\n",
    "    index = 0\n",
    "    id = []\n",
    "    is_it_intron = []\n",
    "    class_of_exon = []\n",
    "    exon_or_intron_number = []\n",
    "    percent_of_homology_for_exons = []\n",
    "    length = []\n",
    "    first_10_nt = []\n",
    "    last_10_nt = []\n",
    "    sequence = []\n",
    "    path = []\n",
    "    \n",
    "    while True: #chyba mimo wszystko zostajemy przy while true bo trudno jest mi wyciagnac liczbe plikow ktore beda lacznie wprowadzane. Tak jest chyba łatwiej\n",
    "        #problem taki wynikajacy ze specyfikacji programu jest taki, że outputowa tabelka bedzie tylko dla jednego organizmu. Potem mozna je jakos polaczyc w calosc\n",
    "        try: \n",
    "            local_homology_percentage = (local_alignment[0].score / distance_between_first_nt_and_gap) #that sign \"[]\" becouse local_alignment's type is list\n",
    "        except IndexError:\n",
    "            if len(seq1) < 100:\n",
    "                print(f\"\\n \\n {alignment[0].id} sequence is too short or sequence has not exons. \\n\")\n",
    "                break\n",
    "            continue\n",
    "        #chwila chwila, czy w tym momencie jest nam wogole potrzebna ta petla while true? bo ona tylko opetla jedna petle ktora sama sie wygasza po spelnieniu warunków\n",
    "        nucleotides = [\"a\", \"t\", \"g\", \"c\", \"A\", \"T\", \"G\", \"C\"]\n",
    "\n",
    "\n",
    "        if len(seq1) > 1:\n",
    "            \n",
    "            if seq1[0] != gaps_signs: #mining data from exons   \n",
    "                index += 1\n",
    "                last_nt_exon_index = seq1.find(gaps_signs) \n",
    "    \n",
    "                if last_nt_exon_index == -1: #it means if it is last exon, if find() don't find any gap signs at the end\n",
    "                    last_nt_exon_index = len(seq1)\n",
    "                    \n",
    "                \n",
    "                alignment_of_exon = pairwise2.align.localxx(seq1[:last_nt_exon_index], seq2[:last_nt_exon_index], one_alignment_only = True)\n",
    "                #print( \"\\n alignment od poczatku nowoprzycietej sekwencji az do ostatniego nukleotydu:\", alignment_of_exon[0])\n",
    "                percent_of_homology = alignment_of_exon[0].score / last_nt_exon_index\n",
    "              \n",
    "                #print(\"zakonczenie analizy eksonow\")\n",
    "                #print(\"score alignmentu = \", alignment_of_exon[0].score)\n",
    "                #print(\"seq1 = \", seq1)\n",
    "                #print(\"alignment = \", alignment_of_exon[0])\n",
    "                \n",
    "\n",
    "                #Creating lists for dictionary for data frame\n",
    "                id.append(alignment[0].id)\n",
    "                exon_or_intron_number.append(index)\n",
    "                is_it_intron.append(False)\n",
    "                percent_of_homology_for_exons.append(round(percent_of_homology * 100, 2))#only for exons\n",
    "                length.append(last_nt_exon_index)\n",
    "                first_10_nt.append(str(seq1[:10]))\n",
    "                last_10_nt.append(str(seq1[last_nt_exon_index-10 : last_nt_exon_index])) \n",
    "                sequence.append(str(seq1[:last_nt_exon_index]))\n",
    "                path.append(str(file))\n",
    "                \n",
    "                if percent_of_homology >= 0.9 and last_nt_exon_index >= min_length_aligned_sequence :\n",
    "                    class_of_exon.append(\"1\")\n",
    "                elif percent_of_homology < 0.9 and percent_of_homology > 0 and last_nt_exon_index < min_length_aligned_sequence:\n",
    "                    class_of_exon.append(\"2\")\n",
    "                else:\n",
    "                    class_of_exon.append(\"3\")\n",
    "\n",
    "                seq1 = seq1[last_nt_exon_index:]\n",
    "                seq2 = seq2[last_nt_exon_index:]\n",
    "                #print(\"zakonczenie analizy eksonow\")\n",
    "\n",
    "                \n",
    "\n",
    "            if len(seq1) > 1 and seq1[0] == \"-\":\n",
    "                index += 1\n",
    "\n",
    "                for i, nukleotide in enumerate(seq1):\n",
    "                    if nukleotide not in nucleotides:\n",
    "                        last_nt_intron_index = i+1\n",
    "                    else:\n",
    "                        break\n",
    "                    \n",
    "\n",
    "        \n",
    "                \n",
    "                id.append(alignment[0].id)\n",
    "                exon_or_intron_number.append(index)\n",
    "                is_it_intron.append(True) \n",
    "                percent_of_homology_for_exons.append(None)\n",
    "                length.append(last_nt_intron_index)\n",
    "                first_10_nt.append(str(seq2[:10]))\n",
    "                last_10_nt.append(str(seq2[last_nt_intron_index-10 : last_nt_intron_index]))\n",
    "                sequence.append(str(seq2[:last_nt_intron_index]))\n",
    "                path.append(str(file))\n",
    "                class_of_exon.append(None)\n",
    "\n",
    "                seq1 = seq1[last_nt_intron_index:]\n",
    "                seq2 = seq2[last_nt_intron_index:]\n",
    "\n",
    "            else:\n",
    "                break\n",
    "                \n",
    "\n",
    "                \n",
    "    dictionary = {\"ID\":id, \"exon_or_intron_number\":exon_or_intron_number, \"is_it_intron\":is_it_intron, \"class_of_exon\":class_of_exon, \"percent_of_homology_for_exons\":percent_of_homology_for_exons, \"length\":length, \"first_10_nt\" : first_10_nt, \"last_10_nt\" : last_10_nt, \"path\":path, \"sequence\" : sequence}\n",
    "      \n",
    "    data_frame = pd.DataFrame(dictionary)\n",
    "    print(\"\\n\\n\\n\\n\\n\", data_frame, \"\\n\\n\\n\\n\\n\")\n",
    "\n",
    "        \n",
    "        \n",
    "\n",
    "\n",
    "#na razie ta funkcja dziala tak: jedna sciezka, jedna ramka. Musze zmienic to tak by wiele sciezek = jedna ramka\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "c4059512-cac4-41e9-86b2-22c0f9f34abd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#cutting_scrap_right(path11,2) \n",
    "#cutting_scrap_right(path12, 2)\n",
    "#cutting_scrap_right(path13, 2)\n",
    "#cutting_scrap_right(path21, 2)\n",
    "#cutting_scrap_right(path22, 2)\n",
    "#cutting_scrap_right(path23, 2)\n",
    "#cutting_scrap_right(path31, 2)\n",
    "#cutting_scrap_right(path32, 2)\n",
    "#cutting_scrap_right(path33, 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "1d0dbc7a-160a-4ae3-96a7-1cbd4df8f3c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                       ID  exon_or_intron_number  is_it_intron  \\\n",
      "0  EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                      1         False   \n",
      "1  EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                      2          True   \n",
      "2  EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                      3         False   \n",
      "3  EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                      4          True   \n",
      "4  EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                      5         False   \n",
      "5  EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                      6          True   \n",
      "6  EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                      7         False   \n",
      "\n",
      "  class_of_exon  percent_of_homology_for_exons  length first_10_nt  \\\n",
      "0             1                          97.92      96  ctacacgtac   \n",
      "1          None                            NaN     102  gcccgccgcc   \n",
      "2             1                         100.00     145  gatcgagctg   \n",
      "3          None                            NaN     238  gaccgggcgg   \n",
      "4             1                         100.00      72  atgtcgccgg   \n",
      "5          None                            NaN     399  cggccagggc   \n",
      "6             1                          97.89     190  cagctgcagg   \n",
      "\n",
      "   last_10_nt                                               path  \\\n",
      "0  ggctgctcct  ./alingment_gen_transkrypt/EG_OG0030307_t1_g1_...   \n",
      "1  cggcgggggc  ./alingment_gen_transkrypt/EG_OG0030307_t1_g1_...   \n",
      "2  ccgctgcttt  ./alingment_gen_transkrypt/EG_OG0030307_t1_g1_...   \n",
      "3  gcccgcggtc  ./alingment_gen_transkrypt/EG_OG0030307_t1_g1_...   \n",
      "4  caccgtggca  ./alingment_gen_transkrypt/EG_OG0030307_t1_g1_...   \n",
      "5  ggcctggggg  ./alingment_gen_transkrypt/EG_OG0030307_t1_g1_...   \n",
      "6  agccccccgg  ./alingment_gen_transkrypt/EG_OG0030307_t1_g1_...   \n",
      "\n",
      "                                            sequence  \n",
      "0  ctacacgtacaagcacgtgaacgtcatggcgccggaggtccaaagc...  \n",
      "1  gcccgccgcccggggttcggtgcccccggctgtcaggccgttgggt...  \n",
      "2  gatcgagctgctgaccggccagccggcgttccccgtctcggaggac...  \n",
      "3  gaccgggcggcctcaccatggcagtgccacatcctgggccaacaga...  \n",
      "4  atgtcgccggacgccctctcgctggccctccgctgcctcgccaagt...  \n",
      "5  cggccagggccgggcaggacaccccacgactcctgcagttcagggg...  \n",
      "6  cagctgcaggcggacccgttcttcgcggacgtggccgtgaagcgcc...   \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                        ID  exon_or_intron_number  \\\n",
      "0   EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                      1   \n",
      "1   EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                      2   \n",
      "2   EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                      3   \n",
      "3   EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                      4   \n",
      "4   EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                      5   \n",
      "5   EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                      6   \n",
      "6   EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                      7   \n",
      "7   EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                      8   \n",
      "8   EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                      9   \n",
      "9   EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                     10   \n",
      "10  EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                     11   \n",
      "11  EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                     12   \n",
      "12  EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                     13   \n",
      "13  EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                     14   \n",
      "14  EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                     15   \n",
      "15  EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                     16   \n",
      "16  EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                     17   \n",
      "17  EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                     18   \n",
      "18  EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                     19   \n",
      "19  EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                     20   \n",
      "20  EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                     21   \n",
      "21  EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                     22   \n",
      "22  EH_OG0030309_t3_g3_EH_PAM1_GP153_OV05                     23   \n",
      "\n",
      "    is_it_intron class_of_exon  percent_of_homology_for_exons  length  \\\n",
      "0          False             1                          97.78      90   \n",
      "1           True          None                            NaN     827   \n",
      "2          False             1                          95.08      61   \n",
      "3           True          None                            NaN     236   \n",
      "4          False             1                         100.00     178   \n",
      "5           True          None                            NaN     343   \n",
      "6          False             1                         100.00     154   \n",
      "7           True          None                            NaN     624   \n",
      "8          False             1                         100.00      84   \n",
      "9           True          None                            NaN     313   \n",
      "10         False             1                         100.00      66   \n",
      "11          True          None                            NaN    2220   \n",
      "12         False             1                         100.00     140   \n",
      "13          True          None                            NaN     546   \n",
      "14         False             1                         100.00      92   \n",
      "15          True          None                            NaN    1603   \n",
      "16         False             1                         100.00      97   \n",
      "17          True          None                            NaN     227   \n",
      "18         False             1                         100.00      76   \n",
      "19          True          None                            NaN     300   \n",
      "20         False             1                         100.00      47   \n",
      "21          True          None                            NaN     751   \n",
      "22         False             1                          99.05     211   \n",
      "\n",
      "   first_10_nt  last_10_nt                                               path  \\\n",
      "0   ttcttctcgg  tctgctgcaa  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "1   ggggccaggc  tgacctgcgc  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "2   ggggccctct  cacctccccg  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "3   accgcccaga  catgtctgca  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "4   accggctgcc  cccaggtccc  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "5   cggccgaggc  gtctcggggg  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "6   cgaggtcatc  gatcggccag  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "7   cagccaatgg  cgattgccac  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "8   caattgcggc  ggattctgaa  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "9   cacccaagtg  cacttgttgc  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "10  catctgatga  tgtctgcccc  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "11  gagcaccccc  cgctggcggg  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "12  gacatcgagc  gggccggcca  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "13  cgcctggggc  tccccaccgc  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "14  cgctcccccg  gtctcgcgct  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "15  ccaccctggg  ggcccagcgg  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "16  ccaagctgcg  agtacggaga  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "17  gagcagggga  ccctggaggc  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "18  gagcttcggg  cgtggagggc  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "19  ctgggcaggg  gccctggggg  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "20  ctggggctgg  agccggaggc  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "21  ggccgtgcca  tcatgatggg  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "22  ggtgctctcc  tataaccagg  ./alingment_gen_transkrypt/EH_OG0030309_t3_g3_...   \n",
      "\n",
      "                                             sequence  \n",
      "0   ttcttctcggagcgccccatgacgcccatcagccacgccgtgttcc...  \n",
      "1   ggggccaggctacgcgggaggggcgtttacatccgaagcacagtat...  \n",
      "2   ggggccctcttcctgaaggcggagctgcccatccgcttggcccacg...  \n",
      "3   accgcccagacattcacctagtgtgcccacaaaacaatgttgcatc...  \n",
      "4   accggctgccccggggcctgggcggcctggcagcgttccagcaggt...  \n",
      "5   cggccgaggctcccatgccccatggtgggtgtgggacagggggcga...  \n",
      "6   cgaggtcatcttccagggctgccagcagctccgcgttcagctgcag...  \n",
      "7   cagccaatggttcgtgtacttccccctacattgaagccccacagtg...  \n",
      "8   caattgcggctccatgcacagtacgtccgaggggagagcccctccg...  \n",
      "9   cacccaagtgccatctgcacaccattaaccaggaaggctctgcttc...  \n",
      "10  catctgatgaaggcgtccatcgcaactgcgaagcgcatggccatcg...  \n",
      "11  gagcaccccccctgtgacactccaacaggtggcttctttaaattca...  \n",
      "12  gacatcgagctcacctgcaagtgtggcacgcccttcatccagatcg...  \n",
      "13  cgcctggggcttggtctaagcaagccaggatccggatgaaagccag...  \n",
      "14  cgctcccccggtgaccatcaccgtggcggacgccgggggcaacgag...  \n",
      "15  ccaccctgggctcgttggctggcccctggccacaagcaaggttggg...  \n",
      "16  ccaagctgcgctacctgtggagctacctgcacgatgatgctcgcct...  \n",
      "17  gagcaggggatggtcttatgccaatttaatattgtcccgtgcgtcg...  \n",
      "18  gagcttcgggctgcccctcgcccgcctgcgggcccggctgttcggc...  \n",
      "19  ctgggcagggctgaccgggacgcccccaaaacaccccccagggtgg...  \n",
      "20    ctggggctggacatgtacgcgtacatcccgaagttcaagccggaggc  \n",
      "21  ggccgtgccacagcttcatggcatcagggtttctggcactggtgct...  \n",
      "22  ggtgctctccaggcccccgttcccgaattgatgtgcgcttttgggc...   \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                        ID  exon_or_intron_number  \\\n",
      "0   EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                      1   \n",
      "1   EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                      2   \n",
      "2   EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                      3   \n",
      "3   EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                      4   \n",
      "4   EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                      5   \n",
      "5   EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                      6   \n",
      "6   EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                      7   \n",
      "7   EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                      8   \n",
      "8   EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                      9   \n",
      "9   EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                     10   \n",
      "10  EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                     11   \n",
      "11  EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                     12   \n",
      "12  EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                     13   \n",
      "13  EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                     14   \n",
      "14  EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                     15   \n",
      "15  EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                     16   \n",
      "16  EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                     17   \n",
      "17  EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                     18   \n",
      "18  EH_OG0030307_t1_g1_EH_PAM1_GP153_OV05                     19   \n",
      "\n",
      "    is_it_intron class_of_exon  percent_of_homology_for_exons  length  \\\n",
      "0          False             1                          100.0     169   \n",
      "1           True          None                            NaN     631   \n",
      "2          False             1                          100.0     107   \n",
      "3           True          None                            NaN     173   \n",
      "4          False             1                          100.0     112   \n",
      "5           True          None                            NaN     308   \n",
      "6          False             1                          100.0      62   \n",
      "7           True          None                            NaN     566   \n",
      "8          False             1                          100.0     154   \n",
      "9           True          None                            NaN     406   \n",
      "10         False             1                          100.0      90   \n",
      "11          True          None                            NaN     232   \n",
      "12         False             1                          100.0     133   \n",
      "13          True          None                            NaN    1814   \n",
      "14         False             1                          100.0      64   \n",
      "15          True          None                            NaN     149   \n",
      "16         False             1                          100.0      46   \n",
      "17          True          None                            NaN     436   \n",
      "18         False             1                          100.0     144   \n",
      "\n",
      "   first_10_nt  last_10_nt                                               path  \\\n",
      "0   gggcgctgtt  gggcccgccc  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "1   cgccccttgg  cgtcctgccg  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "2   caacccattc  cttgaggaac  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "3   ggtatcggcc  ctcttgcgca  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "4   gggtccacca  tacagggggc  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "5   gtccaggctt  gcctgggggt  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "6   gtctttggag  gtgcgagtct  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "7   gctgggtgcc  ccccttggcc  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "8   ggctgtacag  ccatctacac  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "9   gtgtgccccg  tgtcgcacag  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "10  gtacaagcac  ggctgctcct  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "11  gtccaggcct  ccctggcggc  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "12  aattgagctc  ccggtgctct  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "13  ggccgggctg  gtccggggcc  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "14  atgacgccgg  cgccggccca  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "15  ccgtggccag  agaagcctgc  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "16  ccgtggccca  cccctgcggt  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "17  gcccaggctg  gcctgctggg  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "18  gaagcggcag  cgccgagccc  ./alingment_gen_transkrypt/EH_OG0030307_t1_g1_...   \n",
      "\n",
      "                                             sequence  \n",
      "0   gggcgctgttctgtgccacacccataataccaatggcggaggagga...  \n",
      "1   cgccccttgggcgttgcctccgaccccccctccccaaatgttagca...  \n",
      "2   caacccattccatctcgggcctcggcggaaaaatggcgatttcctc...  \n",
      "3   ggtatcggccctcccaccccaatgcggtatttttacttttcctctc...  \n",
      "4   gggtccaccacgcccggggcatgacccacgaccacctcgtgccgta...  \n",
      "5   gtccaggctttgacgcaaggaggcttgtagctaaacaggtggaggg...  \n",
      "6   gtctttggagaccctgctgcggaaacgcaggcgcgggttcgatgag...  \n",
      "7   gctgggtgccgaccgcactccacatgccacggttgtcatgtctatt...  \n",
      "8   ggctgtacagcgtgctgaagggcctgcagcacctccaccagcacgg...  \n",
      "9   gtgtgccccgccccacatccccccgacccaggcaaccggcccccta...  \n",
      "10  gtacaagcacgtcaatgtgatggcgccagaggtccagagcggctcg...  \n",
      "11  gtccaggcctatcgtggtttatatgggacagaccgcgctcgtgcct...  \n",
      "12  aattgagctcctgactggacagcccgccttccccgtctccgaggag...  \n",
      "13  ggccgggctgcgctctggcccccactgtattgagggaggataggga...  \n",
      "14  atgacgccggaagctctcgccctcgccctgcgctgcctcgccccgg...  \n",
      "15  ccgtggccagccttctttttaagggctgctggaccccccccccctg...  \n",
      "16     ccgtggcccagctgctggcggacgacttctttgccgcccctgcggt  \n",
      "17  gcccaggctgaagttggcattgcattggacctcaatcatccccccc...  \n",
      "18  gaagcggcaggcgctgctcggcctgaagcgggcgttgacggaccgc...   \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                        ID  exon_or_intron_number  \\\n",
      "0   EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                      1   \n",
      "1   EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                      2   \n",
      "2   EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                      3   \n",
      "3   EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                      4   \n",
      "4   EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                      5   \n",
      "5   EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                      6   \n",
      "6   EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                      7   \n",
      "7   EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                      8   \n",
      "8   EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                      9   \n",
      "9   EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                     10   \n",
      "10  EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                     11   \n",
      "11  EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                     12   \n",
      "12  EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                     13   \n",
      "13  EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                     14   \n",
      "14  EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                     15   \n",
      "15  EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                     16   \n",
      "16  EL_OG0030309_t3_g3_EL_PAM1_GP153_OV05                     17   \n",
      "\n",
      "    is_it_intron class_of_exon  percent_of_homology_for_exons  length  \\\n",
      "0          False             1                         100.00     227   \n",
      "1           True          None                            NaN     428   \n",
      "2          False             1                         100.00     238   \n",
      "3           True          None                            NaN     316   \n",
      "4          False             1                         100.00     154   \n",
      "5           True          None                            NaN      83   \n",
      "6          False             1                         100.00      83   \n",
      "7           True          None                            NaN     964   \n",
      "8          False             1                         100.00      66   \n",
      "9           True          None                            NaN     292   \n",
      "10         False             1                         100.00     140   \n",
      "11          True          None                            NaN     347   \n",
      "12         False             1                         100.00      95   \n",
      "13          True          None                            NaN     750   \n",
      "14         False             1                         100.00     220   \n",
      "15          True          None                            NaN     768   \n",
      "16         False             1                          99.45     181   \n",
      "\n",
      "   first_10_nt  last_10_nt                                               path  \\\n",
      "0   accgtcaagt  tgttgcaggg  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "1   gcccaggccc  gcctgcacgc  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "2   ggccctcttc  gcacgtccct  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "3   ggacgatgcc  cctcgccatc  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "4   gaggtgattt  gtcggtcagc  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "5   agccagtgcc  ctctgccaat  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "6   agctgcagct  ggattctgaa  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "7   cgcccaagtg  catttgcagt  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "8   cacctgatga  cgtctgtcct  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "9   caccagcgtt  cgctggctgg  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "10  gacattgaac  aggaaggcaa  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "11  tgcctgggtg  gcaccaccac  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "12  tgctgttcct  atctcccggt  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "13  ccaacctgag  ggcccagtgg  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "14  ccaagttgcg  agccagagga  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "15  gtctgtgctg  tcgcagaagt  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "16  ggtgctttct  agtggtgcac  ./alingment_gen_transkrypt/EL_OG0030309_t3_g3_...   \n",
      "\n",
      "                                             sequence  \n",
      "0   accgtcaagtgcctgtcaaccgattccgttggcaaccgccttgagc...  \n",
      "1   gcccaggccccggcctgccattccaggcccggtcatgcttgacttg...  \n",
      "2   ggccctcttcctcaaggcggagctgcccatccgcctggcccactcc...  \n",
      "3   ggacgatgcccctcgtttcccctccccaacccccccacctggccgg...  \n",
      "4   gaggtgattttccagggctgcaagcaacttcgcgctcagctgcacc...  \n",
      "5   agccagtgcctggccactcgatggtgggcccttcttttcctcccgc...  \n",
      "6   agctgcagctccacgcacagtacatcggaggcgaaagcccctcaga...  \n",
      "7   cgcccaagtgcccaagcctccgaagaaaggaaatgacagcaaagaa...  \n",
      "8   cacctgatgaaggcatccattgcaacctgcaagcggatggccatcg...  \n",
      "9   caccagcgttgcttttgcattgggcacattccgagttcgcagcttt...  \n",
      "10  gacattgaactcacctgcaactgtgggacaccgctggtgcaaatcg...  \n",
      "11  tgcctgggtgcgggtttcagcatttgcacaaaagggttgtggaccc...  \n",
      "12  tgctgttcctccgatcactgtgacagtggcggatgccggcaccaat...  \n",
      "13  ccaacctgagcccactgactgcaccggaaaatgatgcagtgtcccc...  \n",
      "14  ccaagttgcgccagctgtggagctatttgcacgacgacgctcacct...  \n",
      "15  gtctgtgctgcagcttgctggtcagcggggggggggggcaggcagg...  \n",
      "16  ggtgctttctggacccgcgtttcaaaagtgacctcctattcgtgtc...   \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " \n",
      " EL_OG0030308_t2_g2_EL_PAM1_GP153_OV05 sequence is too short, sequence has not exons or other fault. \n",
      "\n",
      "\n",
      " \n",
      " EL_OG0030308_t2_g2_EL_PAM1_GP153_OV05 sequence is too short or sequence has not exons. \n",
      "\n",
      "\n",
      " \n",
      " EL_OG0030308_t2_g2_EL_PAM1_GP153_OV05 sequence is too short or sequence has not exons. \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " Empty DataFrame\n",
      "Columns: [ID, exon_or_intron_number, is_it_intron, class_of_exon, percent_of_homology_for_exons, length, first_10_nt, last_10_nt, path, sequence]\n",
      "Index: [] \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                        ID  exon_or_intron_number  \\\n",
      "0   EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                      1   \n",
      "1   EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                      2   \n",
      "2   EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                      3   \n",
      "3   EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                      4   \n",
      "4   EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                      5   \n",
      "5   EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                      6   \n",
      "6   EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                      7   \n",
      "7   EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                      8   \n",
      "8   EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                      9   \n",
      "9   EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     10   \n",
      "10  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     11   \n",
      "11  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     12   \n",
      "12  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     13   \n",
      "13  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     14   \n",
      "14  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     15   \n",
      "15  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     16   \n",
      "16  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     17   \n",
      "17  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     18   \n",
      "18  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     19   \n",
      "19  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     20   \n",
      "20  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     21   \n",
      "21  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     22   \n",
      "22  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     23   \n",
      "23  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     24   \n",
      "24  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     25   \n",
      "25  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     26   \n",
      "26  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     27   \n",
      "27  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     28   \n",
      "28  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     29   \n",
      "29  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     30   \n",
      "30  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     31   \n",
      "31  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     32   \n",
      "32  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     33   \n",
      "33  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     34   \n",
      "34  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     35   \n",
      "35  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     36   \n",
      "36  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     37   \n",
      "37  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     38   \n",
      "38  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     39   \n",
      "39  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     40   \n",
      "40  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     41   \n",
      "41  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     42   \n",
      "42  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     43   \n",
      "43  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     44   \n",
      "44  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     45   \n",
      "45  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     46   \n",
      "46  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     47   \n",
      "47  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     48   \n",
      "48  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     49   \n",
      "49  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     50   \n",
      "50  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     51   \n",
      "51  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     52   \n",
      "52  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     53   \n",
      "53  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     54   \n",
      "54  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     55   \n",
      "55  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     56   \n",
      "56  EG_OG0030308_t2_g2_EG_PAM1_GP153_OV05                     57   \n",
      "\n",
      "    is_it_intron class_of_exon  percent_of_homology_for_exons  length  \\\n",
      "0          False             1                         100.00     150   \n",
      "1           True          None                            NaN     195   \n",
      "2          False             1                         100.00      57   \n",
      "3           True          None                            NaN      49   \n",
      "4          False             1                         100.00      48   \n",
      "5           True          None                            NaN     479   \n",
      "6          False             1                          96.67      60   \n",
      "7           True          None                            NaN     327   \n",
      "8          False             1                          97.62      42   \n",
      "9           True          None                            NaN     473   \n",
      "10         False             1                          98.33      60   \n",
      "11          True          None                            NaN     280   \n",
      "12         False             1                          98.51      67   \n",
      "13          True          None                            NaN    1591   \n",
      "14         False             1                         100.00      40   \n",
      "15          True          None                            NaN     372   \n",
      "16         False             1                          98.55      69   \n",
      "17          True          None                            NaN    1275   \n",
      "18         False             1                         100.00      75   \n",
      "19          True          None                            NaN     197   \n",
      "20         False             1                         100.00      64   \n",
      "21          True          None                            NaN     550   \n",
      "22         False             1                         100.00      42   \n",
      "23          True          None                            NaN     140   \n",
      "24         False             1                         100.00      63   \n",
      "25          True          None                            NaN     310   \n",
      "26         False             1                          99.05     105   \n",
      "27          True          None                            NaN     648   \n",
      "28         False             1                          98.44     128   \n",
      "29          True          None                            NaN     685   \n",
      "30         False             1                         100.00      61   \n",
      "31          True          None                            NaN     889   \n",
      "32         False             1                         100.00     172   \n",
      "33          True          None                            NaN     379   \n",
      "34         False             1                         100.00      67   \n",
      "35          True          None                            NaN      98   \n",
      "36         False             1                         100.00      57   \n",
      "37          True          None                            NaN    1275   \n",
      "38         False             1                         100.00     129   \n",
      "39          True          None                            NaN    1204   \n",
      "40         False             1                         100.00      51   \n",
      "41          True          None                            NaN     660   \n",
      "42         False             1                          99.01     101   \n",
      "43          True          None                            NaN     563   \n",
      "44         False             1                          97.94      97   \n",
      "45          True          None                            NaN     654   \n",
      "46         False             1                         100.00      80   \n",
      "47          True          None                            NaN    1104   \n",
      "48         False             1                         100.00      48   \n",
      "49          True          None                            NaN     280   \n",
      "50         False             1                         100.00      70   \n",
      "51          True          None                            NaN     466   \n",
      "52         False             1                          98.78      82   \n",
      "53          True          None                            NaN     195   \n",
      "54         False             1                         100.00      96   \n",
      "55          True          None                            NaN     388   \n",
      "56         False             1                          99.38     321   \n",
      "\n",
      "   first_10_nt  last_10_nt                                               path  \\\n",
      "0   gtggtgccca  tacagcaact  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "1   ggtgtgcacg  tctctgtaca  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "2   gcaagacatc  atgacttata  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "3   ccaggtgtac  cgggctcctg  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "4   ccaggccaag  acaataccat  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "5   caggtgcgag  ccctcgtggc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "6   cagctacatc  cttccgagtc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "7   agtgtgcagt  tgtatttgca  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "8   ggctagtttt  tgaatgaaga  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "9   caggtgcgct  cgttggtttc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "10  caggagagca  accgacgata  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "11  gtgtgtcctt  tttttggcag  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "12  ccagttgatt  ggattccgtg  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "13  ggtgtgcttt  ccttgcccca  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "14  ggcaagctgg  accgcggcca  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "15  caggtgtgga  cattattgtg  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "16  cagcggtgaa  atataccaag  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "17  aggtgtgtgt  cccgtccccc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "18  aggtcaaggg  aacttcgggc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "19  ggtgtgcccc  ctcccctgca  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "20  ggctgttttt  tacaggagct  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "21  ggtattcgaa  gtttcccaca  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "22  gcccgaggag  ggtcattgaa  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "23  aggtgtggag  tccgcgccgc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "24  agtcatcgcg  acgtcacggg  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "25  ggtgtggggc  ctcaaccgta  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "26  ggtgtatctc  agaagtcata  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "27  tgcaggtcgc  ctcttgcgcc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "28  tgcagagtac  gacataaaga  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "29  aggtgtggct  ctgcttgctt  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "30  aggtgctgtt  tcaatattcg  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "31  aggtgtagca  cgtcgggcgc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "32  aggggacatc  ggccctcctc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "33  ggtgggcccc  tgccgacgca  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "34  gaggccggag  aagatggaag  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "35  aggtgtgccc  tgtgtcgggc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "36  agacggaccg  tccatcgcga  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "37  aggtgtggcg  actgttgctc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "38  aggcgggtat  ttgctctccc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "39  ggcatgtccc  tttccctgca  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "40  ggttcgacat  gacagtgacc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "41  ggtgcgcttg  cccttccaca  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "42  ggaccctggc  caacctactg  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "43  aggtttgtgt  acccatcccc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "44  agagccatga  cgggcggcca  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "45  aggtgcgcat  ccacgcccac  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "46  aggacgacca  ccaggcccat  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "47  gtgtgttggg  tcccttccag  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "48  gcacggctcc  cgtcgaagag  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "49  gcaggtcccg  cattcgcttc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "50  gcagagcggc  caggactccc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "51  ggtccgtccc  cccctcggca  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "52  ggtttcgcgg  acagcgtcaa  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "53  gtgtgcgtcg  gccctcccag  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "54  gctggacgcg  acgtctgggc  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "55  ggtgtgggtc  gtctccgaaa  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "56  ggtggcccac  aagcagagat  ./alingment_gen_transkrypt/EG_OG0030308_t2_g2_...   \n",
      "\n",
      "                                             sequence  \n",
      "0   gtggtgcccatcggcaaggaccaggatggcgtagccggggatggcc...  \n",
      "1   ggtgtgcacgtgtgtgtgggcgcagatgtgctcgcttttttgattg...  \n",
      "2   gcaagacatctctgaccggcggacgaagatattcgacctgtccctg...  \n",
      "3   ccaggtgtacccccggctttatggcttatggctcacattcgggctcctg  \n",
      "4    ccaggccaaggagaacgaattggcagacaagatccggaacaataccat  \n",
      "5   caggtgcgagccaccccagtgcagatgtggctttttcgaggaccaa...  \n",
      "6   cagctacatcgagcagctctccagagtcgtcgatgacaacctgcca...  \n",
      "7   agtgtgcagtgtaccttctgtgtgttttgtgtgatgtgtttctggg...  \n",
      "8          ggctagttttgagcgcgacactttagacgtgctgaatgaaga  \n",
      "9   caggtgcgctgctcattttttcatcgagtctcaggttcatcgtcca...  \n",
      "10  caggagagcaaatggcgatgatgcaggaaaaggcttgccagcggca...  \n",
      "11  gtgtgtcctttcatgctgtccatcaatgctgaattcatgctgtgaa...  \n",
      "12  ccagttgattgtcacgtatgatgggcagaagcgactcccactccgc...  \n",
      "13  ggtgtgctttgtcgtcacccgcagcaagtcaaactgccattggttg...  \n",
      "14           ggcaagctggtgtgcatccaggggattgtcaccgcggcca  \n",
      "15  caggtgtggatgctctgttccttgaacctttggtcgcttggcccat...  \n",
      "16  cagcggtgaagccaaagattcaatttgctgcatacatatgtgagca...  \n",
      "17  aggtgtgtgtgtgtgtgtgtgtgtgtgtgtgtgtgtgtgtgtggtg...  \n",
      "18  aggtcaagggcgacgcctacacccccgagctgtcttgcagctccga...  \n",
      "19  ggtgtgccccccgcctcttgccttgtgtgttgggcacgtggacaga...  \n",
      "20  ggctgtttttccagccacgcggcagcaagttcaccaagttccagga...  \n",
      "21  ggtattcgaagttggatttctggccgagagtgtttctgatttcttt...  \n",
      "22         gcccgaggaggtgccgaaggggtccataccgcggtcattgaa  \n",
      "23  aggtgtggagccccgcggcggccgaactggacccgttgggccgtgt...  \n",
      "24  agtcatcgcggaggagggcgcgtcctcgcagggcacgccaggggac...  \n",
      "25  ggtgtggggctcgcccccccccctcctctctcagttcccttctctg...  \n",
      "26  ggtgtatctccccgagtcggcccaggggttcaaggcactgatggcc...  \n",
      "27  tgcaggtcgccatcacctcttttttggtggcgtgtttgtgtgtgtg...  \n",
      "28  tgcagagtacacggcgtcccggaagctgcaggaggaggtggaggcg...  \n",
      "29  aggtgtggctgaggagtgggtgtgatgacgtgcagcggggccacag...  \n",
      "30  aggtgctgttgttgcagcttgttggcgctggagcgaagtcattgtc...  \n",
      "31  aggtgtagcatccggcccggtgtaggtcaccagttggctgtagatg...  \n",
      "32  aggggacatcaacgtctgcctgatgggggaccccggcgtcgccaag...  \n",
      "33  ggtgggcccctgccccccttgacggtcgcagccgccttcgccgatg...  \n",
      "34  gaggccggagcactggttctggcagacaatggcatttgctgcatcg...  \n",
      "35  aggtgtgccctggcagctgtcctgtgttggcggtactgttggcaca...  \n",
      "36  agacggaccggacctccatccacgaggtgatggagcagcagacggt...  \n",
      "37  aggtgtggcgctccctctgtcgtctgtccttggcgccactggcaaa...  \n",
      "38  aggcgggtatcgttacgtctctaaatgcccgggtctcggttctagc...  \n",
      "39  ggcatgtccctatcatggcccatcccaggaggaattggtttttgag...  \n",
      "40  ggttcgacatcctgtggctgctcctggatgagcctgacgcggacag...  \n",
      "41  ggtgcgcttggcctcctcttttggtgttttttttggaggtgcgccc...  \n",
      "42  ggaccctggcgaactttgttgccaagaatcacctcatcacacagga...  \n",
      "43  aggtttgtgttggcgcctgtgaccgtcgtggggtgcccccacccac...  \n",
      "44  agagccatgatcgggcaggcgaaggagtattgcccagcggtgccga...  \n",
      "45  aggtgcgcatgttgtccgctgtttgtcccaaggatggtacggcgct...  \n",
      "46  aggacgaccagatcgccgacgccaccacgtgggtcaccccccggac...  \n",
      "47  gtgtgttgggcggtgttggagttggctttttgttgcgttgcggcag...  \n",
      "48   gcacggctccggcggtccaattccatcaacgaggaggacgtcgaagag  \n",
      "49  gcaggtcccgccccgccgcaactcctcccatcgctgccatccttct...  \n",
      "50  gcagagcggctgatgaagcagtccaagcgcagcctgattgacgccc...  \n",
      "51  ggtccgtccccgcctcgtcccaccacccttgcccccccccccagcc...  \n",
      "52  ggtttcgcggcggggatttctccgcccggatcttcgcgatcatcct...  \n",
      "53  gtgtgcgtcgggttcggggcttgccacacgtcccctctctcccatt...  \n",
      "54  gctggacgcggtgcgacagcagctcaaggtgcgggggttccgggag...  \n",
      "55  ggtgtgggtccttccccttccccctatccttttttgggtgtaccgg...  \n",
      "56  ggtggcccacgactactccgaggtgcggttctacacctaaccgccc...   \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                       ID  exon_or_intron_number  is_it_intron  \\\n",
      "0  EL_OG0030307_t1_g1_EL_PAM1_GP153_OV05                      1         False   \n",
      "1  EL_OG0030307_t1_g1_EL_PAM1_GP153_OV05                      2          True   \n",
      "2  EL_OG0030307_t1_g1_EL_PAM1_GP153_OV05                      3         False   \n",
      "3  EL_OG0030307_t1_g1_EL_PAM1_GP153_OV05                      4          True   \n",
      "4  EL_OG0030307_t1_g1_EL_PAM1_GP153_OV05                      5         False   \n",
      "\n",
      "  class_of_exon  percent_of_homology_for_exons  length first_10_nt  \\\n",
      "0             1                          100.0     136  gattgaactg   \n",
      "1          None                            NaN    2441  ggccggacca   \n",
      "2             1                          100.0      68  atgacaccag   \n",
      "3          None                            NaN     313  cccacagcca   \n",
      "4             1                          100.0     254  cccacagctg   \n",
      "\n",
      "   last_10_nt                                               path  \\\n",
      "0  tcgctgcttt  ./alingment_gen_transkrypt/EL_OG0030307_t1_g1_...   \n",
      "1  gtccgtggcc  ./alingment_gen_transkrypt/EL_OG0030307_t1_g1_...   \n",
      "2  ggcccactgt  ./alingment_gen_transkrypt/EL_OG0030307_t1_g1_...   \n",
      "3  aaccgccctg  ./alingment_gen_transkrypt/EL_OG0030307_t1_g1_...   \n",
      "4  cctgcccccc  ./alingment_gen_transkrypt/EL_OG0030307_t1_g1_...   \n",
      "\n",
      "                                            sequence  \n",
      "0  gattgaactgctgacgggacagcccgcattcccggcctcggaggac...  \n",
      "1  ggccggaccaccttctggtgctcccacaactgtctttcaggatgga...  \n",
      "2  atgacaccagatgccctcgccctcgccctttgctgcctgcgcaaat...  \n",
      "3  cccacagccagggaggtcctcagggacaggattcttttccccctct...  \n",
      "4  cccacagctgttggctgaccccttcttctcggacccggcggtgaag...   \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                       ID  exon_or_intron_number  is_it_intron  \\\n",
      "0  EH_OG0030308_t2_g2_EH_PAM1_GP153_OV05                      1         False   \n",
      "1  EH_OG0030308_t2_g2_EH_PAM1_GP153_OV05                      2          True   \n",
      "2  EH_OG0030308_t2_g2_EH_PAM1_GP153_OV05                      3         False   \n",
      "3  EH_OG0030308_t2_g2_EH_PAM1_GP153_OV05                      4          True   \n",
      "4  EH_OG0030308_t2_g2_EH_PAM1_GP153_OV05                      5         False   \n",
      "\n",
      "  class_of_exon  percent_of_homology_for_exons  length first_10_nt  \\\n",
      "0             1                          99.32     147  cgaacttcat   \n",
      "1          None                            NaN      64  ggtgtgcctt   \n",
      "2             1                         100.00      58  gcaagagatc   \n",
      "3          None                            NaN      52  caggtagctc   \n",
      "4             1                         100.00      48  caggcaaaag   \n",
      "\n",
      "   last_10_nt                                               path  \\\n",
      "0  tgaggcaatt  ./alingment_gen_transkrypt/EH_OG0030308_t2_g2_...   \n",
      "1  tttgcatcca  ./alingment_gen_transkrypt/EH_OG0030308_t2_g2_...   \n",
      "2  tgatctgtac  ./alingment_gen_transkrypt/EH_OG0030308_t2_g2_...   \n",
      "3  tctctttggt  ./alingment_gen_transkrypt/EH_OG0030308_t2_g2_...   \n",
      "4  aaacaccatt  ./alingment_gen_transkrypt/EH_OG0030308_t2_g2_...   \n",
      "\n",
      "                                            sequence  \n",
      "0  cgaacttcatgcgaagaagcataatgtgctaacaaggatggccgaa...  \n",
      "1  ggtgtgccttttgcattgtgccttttgcattggacgtgtcgctgcc...  \n",
      "2  gcaagagatctccgacaggaagacaagaatatttgaattatcgttg...  \n",
      "3  caggtagctccacgcttgctttataatgtggtccaatggcattctc...  \n",
      "4   caggcaaaagaagatgaattggccgacaagattcggacaaacaccatt   \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                        ID  exon_or_intron_number  \\\n",
      "0   EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                      1   \n",
      "1   EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                      2   \n",
      "2   EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                      3   \n",
      "3   EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                      4   \n",
      "4   EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                      5   \n",
      "5   EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                      6   \n",
      "6   EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                      7   \n",
      "7   EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                      8   \n",
      "8   EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                      9   \n",
      "9   EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                     10   \n",
      "10  EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                     11   \n",
      "11  EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                     12   \n",
      "12  EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                     13   \n",
      "13  EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                     14   \n",
      "14  EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                     15   \n",
      "\n",
      "    is_it_intron class_of_exon  percent_of_homology_for_exons  length  \\\n",
      "0          False             1                          98.86      88   \n",
      "1           True          None                            NaN     132   \n",
      "2          False             1                         100.00      57   \n",
      "3           True          None                            NaN    1299   \n",
      "4          False             1                         100.00      52   \n",
      "5           True          None                            NaN    2440   \n",
      "6          False             1                          98.08      52   \n",
      "7           True          None                            NaN     635   \n",
      "8          False             1                          99.24     132   \n",
      "9           True          None                            NaN     347   \n",
      "10         False             1                          99.35     154   \n",
      "11          True          None                            NaN     356   \n",
      "12         False             1                         100.00      84   \n",
      "13          True          None                            NaN    1669   \n",
      "14         False             1                         100.00      67   \n",
      "\n",
      "   first_10_nt  last_10_nt                                               path  \\\n",
      "0   tggcagcggt  ccatcagcca  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "1   cgccgtgcaa  acgtgcttgg  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "2   cgccgtgttc  tgctgcaggg  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "3   ggccaggccg  cctgtggcac  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "4   agccctcttc  ctccgcgcac  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "5   ctccctgtaa  accaacctta  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "6   ctccctgacc  ttccagcagg  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "7   gtgcgggagg  atgggcaacc  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "8   tgcgggagat  aagcgggtcc  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "9   ctggtcgagg  ggtctcgacg  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "10  ctgagatcgt  tgatcggcca  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "11  gcgcccagtg  gtcactgcgg  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "12  gcaattgcgg  tggattctga  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "13  gcgccgaagt  gcgcttgcac  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "14  gcatctgatg  tgtatgccct  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "\n",
      "                                             sequence  \n",
      "0   tggcagcggtgcgggagcgggcggagtacctcgacgtgatggagga...  \n",
      "1   cgccgtgcaagcacctacataatgactcggcgtgactgcgcgtgac...  \n",
      "2   cgccgtgttccaccgcttcgccttgcaccagacccagaaggcggcg...  \n",
      "3   ggccaggccgtggcgcaggccttcgcgctggccctcggcagcccct...  \n",
      "4   agccctcttcctgaaggcggagctgcccatccgcctcgcccactcc...  \n",
      "5   ctccctgtaaggttggtaagactagtcgactagtcgattattcatt...  \n",
      "6   ctccctgaccgcctgccccgcggcctcagcggcctcgccgcgttcc...  \n",
      "7   gtgcgggaggccaggctgcccttgccggcaaccaacgaccctatcc...  \n",
      "8   tgcgggagatgttcacggactccttccgggagctgcgcgcgttcca...  \n",
      "9   ctggtcgaggcctgacacaaaatcgacacaccctctccaaaggggt...  \n",
      "10  ctgagatcgtcttccagggctgccagcagctccgcacgcagctcca...  \n",
      "11  gcgcccagtgccttgcggggcccactggcagtgctgcccaggaacc...  \n",
      "12  gcaattgcggctccatgcgcagtacgtgcggggggagagccacgcg...  \n",
      "13  gcgccgaagtgccatttgttggaacctgaggattttccttaggggt...  \n",
      "14  gcatctgatgaaggcgtccatagcaaccgccaaacggatggccctc...   \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "directory = './alingment_gen_transkrypt'\n",
    "\n",
    "for filename in os.listdir(directory):\n",
    "    f = os.path.join(directory, filename)\n",
    "    if not os.path.isfile(f):\n",
    "        continue\n",
    "    if f.endswith(\".aln\"):\n",
    "        #print(\"sciezka: \\t\", f)\n",
    "        cutting_scrap_right(f, 2)\n",
    "        print(\"\\n\\n\\n\\n\")\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae86ba7e-ac79-44f8-a430-738225aa577f",
   "metadata": {},
   "source": [
    "Pierwsza decyzja w przypadku poprawiania kodu na jego prawy odpowiednik (reversed):\n",
    "- nazwy zmiennych które jawnie odnoszą się do pozycji, np. lewe coś tam, pierwszy nukleotyd zostały zmienione na ich właściwe odpowiednki\n",
    "- Pozostałe elementy zostaja spojne - alignment to wciaz zmienna alignment pojawiajaca sie w obu fragmentach kodu.\n",
    "\n",
    "Druga decyzja:\n",
    "- kod bedzie mial strukture:\n",
    "      - przycinanie lewej strony\n",
    "      - przycinanie prawej strony\n",
    "      - wysylka do excela\n",
    "  - i żaden z tych fragmentow nie "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffc6c601-63fc-4f1c-9e5f-a148bc88acbf",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
