{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f037dd59-8d78-4bbc-b1f8-f878afe1f342",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/norbert/miniconda3/lib/python3.11/site-packages/Bio/pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from Bio import AlignIO, SeqIO, Seq, pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e7f5ae9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "min_length_aligned_sequence = 30 #Minimal lenght of sequence which could be anÂ exon\n",
    "extreme_homology = 0.97 #percentage of homology of sequence, treshold #I assume two faulty aligned nucleotides per 100 (98%) and one more nt because sometimes latest nt can move from end of one sequence to beginning next sequence\n",
    "middle_homology = 0.9 #used for classification of internal exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e6d4c5ee-799f-427e-839a-f95917b6a430",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " \n",
      " EL_OG0030308_t2_g2_EL_PAM1_GP153_OV05 sequence is too short, sequence has not exons or other fault. \n",
      "\n",
      "\n",
      " \n",
      " EL_OG0030308_t2_g2_EL_PAM1_GP153_OV05 sequence is too short or sequence has not exons. \n",
      "\n",
      "\n",
      " \n",
      " EL_OG0030308_t2_g2_EL_PAM1_GP153_OV05 sequence is too short or sequence has not exons. \n",
      "\n",
      "                                        ID  exon_or_intron_number  \\\n",
      "1    EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                    1.0   \n",
      "2    EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                    2.0   \n",
      "3    EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                    3.0   \n",
      "4    EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                    4.0   \n",
      "5    EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                    5.0   \n",
      "..                                     ...                    ...   \n",
      "144  EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                   11.0   \n",
      "145  EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                   12.0   \n",
      "146  EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                   13.0   \n",
      "147  EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                   14.0   \n",
      "148  EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                   15.0   \n",
      "\n",
      "    is_it_intron class_of_exon  percent_of_homology_for_exons  length  \\\n",
      "1            0.0             1                          97.92    96.0   \n",
      "2            1.0          None                            NaN   102.0   \n",
      "3            0.0             1                         100.00   145.0   \n",
      "4            1.0          None                            NaN   238.0   \n",
      "5            0.0             1                         100.00    72.0   \n",
      "..           ...           ...                            ...     ...   \n",
      "144          0.0             1                          99.35   154.0   \n",
      "145          1.0          None                            NaN   356.0   \n",
      "146          0.0             1                         100.00    84.0   \n",
      "147          1.0          None                            NaN  1669.0   \n",
      "148          0.0             1                         100.00    67.0   \n",
      "\n",
      "    first_10_nt  last_10_nt  \\\n",
      "1    ctacacgtac  ggctgctcct   \n",
      "2    gcccgccgcc  cggcgggggc   \n",
      "3    gatcgagctg  ccgctgcttt   \n",
      "4    gaccgggcgg  gcccgcggtc   \n",
      "5    atgtcgccgg  caccgtggca   \n",
      "..          ...         ...   \n",
      "144  ctgagatcgt  tgatcggcca   \n",
      "145  gcgcccagtg  gtcactgcgg   \n",
      "146  gcaattgcgg  tggattctga   \n",
      "147  gcgccgaagt  gcgcttgcac   \n",
      "148  gcatctgatg  tgtatgccct   \n",
      "\n",
      "                                                  path  \\\n",
      "1    ./alingment_gen_transkrypt/EG_OG0030307_t1_g1_...   \n",
      "2    ./alingment_gen_transkrypt/EG_OG0030307_t1_g1_...   \n",
      "3    ./alingment_gen_transkrypt/EG_OG0030307_t1_g1_...   \n",
      "4    ./alingment_gen_transkrypt/EG_OG0030307_t1_g1_...   \n",
      "5    ./alingment_gen_transkrypt/EG_OG0030307_t1_g1_...   \n",
      "..                                                 ...   \n",
      "144  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "145  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "146  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "147  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "148  ./alingment_gen_transkrypt/EG_OG0030309_t3_g3_...   \n",
      "\n",
      "                                              sequence  \n",
      "1    ctacacgtacaagcacgtgaacgtcatggcgccggaggtccaaagc...  \n",
      "2    gcccgccgcccggggttcggtgcccccggctgtcaggccgttgggt...  \n",
      "3    gatcgagctgctgaccggccagccggcgttccccgtctcggaggac...  \n",
      "4    gaccgggcggcctcaccatggcagtgccacatcctgggccaacaga...  \n",
      "5    atgtcgccggacgccctctcgctggccctccgctgcctcgccaagt...  \n",
      "..                                                 ...  \n",
      "144  ctgagatcgtcttccagggctgccagcagctccgcacgcagctcca...  \n",
      "145  gcgcccagtgccttgcggggcccactggcagtgctgcccaggaacc...  \n",
      "146  gcaattgcggctccatgcgcagtacgtgcggggggagagccacgcg...  \n",
      "147  gcgccgaagtgccatttgttggaacctgaggattttccttaggggt...  \n",
      "148  gcatctgatgaaggcgtccatagcaaccgccaaacggatggccctc...  \n",
      "\n",
      "[148 rows x 10 columns]\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "File already exists. Do you want overwrite? y/n:  25\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Clarify your answer\n"
     ]
    }
   ],
   "source": [
    "def cutting_scrap_right(directory, acceptable_gap_length):\n",
    "    data_frames = []\n",
    "    for filename in os.listdir(directory):\n",
    "        file = os.path.join(directory, filename)\n",
    "        if not os.path.isfile(file):\n",
    "            continue\n",
    "        if file.endswith(\".aln\"):\n",
    "            gaps_signs = \"-\" * acceptable_gap_length #maximum length of gaps in sequence in one exon's sequence\n",
    "    \n",
    "            alignment = AlignIO.read(file, \"clustal\")\n",
    "    \n",
    "        \n",
    "        \n",
    "            while True:\n",
    "                first_nucleotides_pair = alignment[:, 0]#first nucleotides in both strands as variable\n",
    "                if \"-\" in first_nucleotides_pair: #deleting mismatches at the start \n",
    "                    alignment = alignment[:, 1:]\n",
    "                    #print(f\"alignment after cut:\\n{alignment}\") #chekpoint\n",
    "                    continue\n",
    "    \n",
    "            \n",
    "        \n",
    "                #changing extension\n",
    "                seq1 = alignment[0].seq #transcript sequence\n",
    "                seq2 = alignment[1].seq #gene sequence\n",
    "        \n",
    "                \n",
    "                #Counting distance:\n",
    "                first_gaps_position_seq1 = seq1.find(gaps_signs) #first gap inddex.\n",
    "                first_gaps_position_seq2 = seq2.find(gaps_signs) \n",
    "                #print(f\"first gap position in first sequence:{first_gaps_position_seq1}, in 2nd sequence: {first_gaps_position_seq2}\") #checkpoint\n",
    "                \n",
    "                if first_gaps_position_seq2 < 0: #avoiding faults when find() function returns \"-1\" as finding nothing\n",
    "                    distance_between_first_nt_and_gap = first_gaps_position_seq1\n",
    "                else:\n",
    "                    distance_between_first_nt_and_gap = min(first_gaps_position_seq1, first_gaps_position_seq2)\n",
    "        \n",
    "                \n",
    "                local_alignment = pairwise2.align.localxx(seq1[:distance_between_first_nt_and_gap], seq2[:distance_between_first_nt_and_gap], \\\n",
    "                                                            one_alignment_only = True)\n",
    "                #\"xx\"  means no gap penalty while opening gaps or longering them and no penalties for mismatch. Just pure score of alignment to count homology\n",
    "                #print(\"local_alignment:\", local_alignment) #checkpoint\n",
    "               \n",
    "                try: \n",
    "                    local_homology_percentage = (local_alignment[0].score / distance_between_first_nt_and_gap) \n",
    "                except IndexError:\n",
    "                    if len(seq1) < 100:\n",
    "                        print(f\"\\n \\n {alignment[0].id} sequence is too short, sequence has not exons or other fault. \\n\")\n",
    "                        break\n",
    "                    continue\n",
    "                \n",
    "                \n",
    "                if local_homology_percentage <= extreme_homology:\n",
    "                    #print(f\"\\n Too low score of homology. \\n alignment: {alignment[:, :50]} \\n score: {local_alignment[0].score}, distance: {distance_between_first_nt_and_gap}, percent: {round(local_homology_percentage, 2)*100}%, \\n {local_alignment}\")\n",
    "                    alignment = alignment[:, distance_between_first_nt_and_gap:] \n",
    "                    \n",
    "                else:\n",
    "                    if distance_between_first_nt_and_gap <= min_length_aligned_sequence:\n",
    "                        #print(f\"\\n alignment is too short. \\n distance between sequences {distance_between_first_nt_and_gap}, \\n alignment: {alignment}, \\n score: {local_alignment[0].score}, distance: {distance_between_first_nt_and_gap}, procentowo: {round(local_homology_percentage, 2)*100}%, \\n {local_alignment}\")\n",
    "                        alignment = alignment[:, distance_between_first_nt_and_gap:]\n",
    "                    else:\n",
    "                        #print(f\"Sequence {alignment[0].id} cut properly from LEFT side, \\n distance between sequences: {distance_between_first_nt_and_gap}\")\n",
    "                        #alignment: {alignment[:, :50]},\n",
    "                        break\n",
    "                \n",
    "        \n",
    "            \n",
    "        ######################### TERAZ OD PRAWEJ DO LEWEJ #######################  \n",
    "        \n",
    "            \n",
    "            while True:\n",
    "                last_nucleotides_pair = alignment[:, -1]#last nucleotides in both strands as variable\n",
    "                if \"-\" in last_nucleotides_pair: #deleting mismatches at the end\n",
    "                    alignment = alignment[:, :-1]\n",
    "                    #print(f\"alignment check:\\n{alignment[:, -50:-1]}\") checkpoint\n",
    "                    continue\n",
    "                        \n",
    "                            \n",
    "                seq1 = alignment[0].seq #transcript sequence\n",
    "                seq2 = alignment[1].seq #gene sequence\n",
    "                \n",
    "                last_gaps_position_seq1 = seq1.rfind(gaps_signs)  #Counting distance: #getting index of first gap. \n",
    "                last_gaps_position_seq1 = last_gaps_position_seq1 + acceptable_gap_length\n",
    "                \n",
    "                last_gaps_position_seq2 = seq2.rfind(gaps_signs)  \n",
    "                last_gaps_position_seq2 = last_gaps_position_seq2 + acceptable_gap_length #now we have to add length of gap to index, because python starts counting from 0 nt, and we just change direction of counting.\n",
    "                #print(\"seq1:\", last_gaps_position_seq1, \"seq2:\", last_gaps_position_seq2) checkpoint\n",
    "                \n",
    "                \n",
    "                if last_gaps_position_seq2 < 0: #it means, if there is no gaps in second sequention, gap position is gap position in seq1. Additional, rfind() function return \"-1\" if it won't find any gaps in seq\n",
    "                    distance_between_first_nt_and_gap = (len(seq1) - last_gaps_position_seq1)\n",
    "                    distance_between_first_nt_and_gap = distance_between_first_nt_and_gap - 1 #correction to counting from 0th index\n",
    "                    #print(\"pierwszy\", distance_between_first_nt_and_gap)\n",
    "                else:\n",
    "                    distance_between_first_nt_and_gap = (len(seq1) - max(last_gaps_position_seq1, last_gaps_position_seq2))\n",
    "                    #print(\"drugi\", distance_between_first_nt_and_gap)\n",
    "                    \n",
    "                    \n",
    "                    local_alignment = pairwise2.align.localxx(seq1[-distance_between_first_nt_and_gap:], seq2[-distance_between_first_nt_and_gap:], \\\n",
    "                                                            one_alignment_only = True)\n",
    "                    #\"xx\" means no gap penalty while opening gaps or longering them and no penalties for mismatch. Just pure score of alignment\n",
    "                #print(\"local_alignment:\", local_alignment) #checkpoint\n",
    "                \n",
    "                \n",
    "        \n",
    "                try:\n",
    "                    local_homology_percentage = (local_alignment[0].score / distance_between_first_nt_and_gap) \n",
    "                except IndexError:\n",
    "                    if len(seq1) < 100:\n",
    "                        print(f\"\\n \\n {alignment[0].id} sequence is too short or sequence has not exons. \\n\")\n",
    "                        break\n",
    "                    continue\n",
    "        \n",
    "                \n",
    "                if local_homology_percentage < extreme_homology:\n",
    "                    #print(f\"\\n Too low homology! \\n alignment: {alignment[:, -50:]} \\n score: {local_alignment[0].score}, \\distance: {distance_between_first_nt_and_gap}, %%%: {round(local_homology_percentage, 2)*100}%, \\n {local_alignment}\")\n",
    "                    alignment = alignment[:, :-distance_between_first_nt_and_gap] \n",
    "                    \n",
    "                else:\n",
    "                    if distance_between_first_nt_and_gap < min_length_aligned_sequence:\n",
    "                        #print(f\"\\n Alignment is too short!. \\n distance: {distance_between_first_nt_and_gap}, \\n alignment: {alignment}, \\n score: {local_alignment[0].score}, distance: {distance_between_first_nt_and_gap}, %%%: {round(local_homology_percentage, 2)*100}%, \\n {local_alignment}\")\n",
    "                        alignment = alignment[:, :-distance_between_first_nt_and_gap]\n",
    "                    else:\n",
    "                        #print(f\"Sequence: {alignment[0].id} cut properly from RIGHT side\")\n",
    "                        #print(f\"\"\"\\n \\n \\n victory prawo! \\n distance between sequences {distance_between_first_nt_and_gap},\n",
    "                        #alignment: {alignment[:, -50:]}, \n",
    "                        #score: {local_alignment[0].score}, distance: {distance_between_first_nt_and_gap}, \n",
    "                        #percent: {round(local_homology_percentage, 4)*100}%, \\n \\n local alignment to confirm: \\n{local_alignment}\"\"\")\n",
    "                        break\n",
    "        \n",
    "        \n",
    "        \n",
    "            \n",
    "            nucleotides = [\"a\", \"t\", \"g\", \"c\", \"A\", \"T\", \"G\", \"C\"]\n",
    "            index = 0\n",
    "            id = []\n",
    "            is_it_intron = []\n",
    "            class_of_exon = []\n",
    "            exon_or_intron_number = []\n",
    "            percent_of_homology_for_exons = []\n",
    "            length = []\n",
    "            first_10_nt = []\n",
    "            last_10_nt = []\n",
    "            sequence = []\n",
    "            path = []\n",
    "            \n",
    "            while True:\n",
    "                try: \n",
    "                    local_homology_percentage = (local_alignment[0].score / distance_between_first_nt_and_gap) #that sign \"[]\" becouse local_alignment's type is list\n",
    "                except IndexError:\n",
    "                    if len(seq1) < 100:\n",
    "                        print(f\"\\n \\n {alignment[0].id} sequence is too short or sequence has not exons. \\n\")\n",
    "                        break\n",
    "                    continue\n",
    "                \n",
    "                if len(seq1) > 1:\n",
    "                        \n",
    "                    if seq1[0] != gaps_signs: #mining data from exons   \n",
    "                        index += 1\n",
    "                        last_nt_exon_index = seq1.find(gaps_signs) \n",
    "                \n",
    "                        if last_nt_exon_index == -1: #it means if it is last exon, if find() don't find any gap signs at the end\n",
    "                            last_nt_exon_index = len(seq1)\n",
    "                                \n",
    "                        alignment_of_exon = pairwise2.align.localxx(seq1[:last_nt_exon_index], seq2[:last_nt_exon_index], one_alignment_only = True)\n",
    "                        #print( \"\\n alignment since start to the end of exon:\", alignment_of_exon[0])\n",
    "                        percent_of_homology = alignment_of_exon[0].score / last_nt_exon_index\n",
    "                          \n",
    "                        #print(\"exon analising finished\")\n",
    "                        #print(f\"score alignmentu = , {alignment_of_exon[0].score}\\n seq1 = {seq1}, \\n alignment = , {alignment_of_exon[0]}\")\n",
    "                            \n",
    "            \n",
    "\n",
    "                    id.append(alignment[0].id)\n",
    "                    exon_or_intron_number.append(index)\n",
    "                    is_it_intron.append(False)\n",
    "                    percent_of_homology_for_exons.append(round(percent_of_homology * 100, 2))\n",
    "                    length.append(last_nt_exon_index)\n",
    "                    first_10_nt.append(str(seq1[:10]))\n",
    "                    last_10_nt.append(str(seq1[last_nt_exon_index-10 : last_nt_exon_index])) \n",
    "                    sequence.append(str(seq1[:last_nt_exon_index]))\n",
    "                    path.append(str(file))\n",
    "                            \n",
    "                    if percent_of_homology >= 0.9 and last_nt_exon_index >= min_length_aligned_sequence :\n",
    "                        class_of_exon.append(\"1\")\n",
    "                    elif percent_of_homology < 0.9 and percent_of_homology > 0 and last_nt_exon_index < min_length_aligned_sequence:\n",
    "                        class_of_exon.append(\"2\")\n",
    "                    else:\n",
    "                        class_of_exon.append(\"3\")\n",
    "            \n",
    "                    seq1 = seq1[last_nt_exon_index:]\n",
    "                    seq2 = seq2[last_nt_exon_index:]\n",
    "                    #print(\"analising exon finished\")\n",
    "            \n",
    "                            \n",
    "            \n",
    "                if len(seq1) > 1 and seq1[0] == \"-\":\n",
    "                    index += 1\n",
    "            \n",
    "                    for i, nukleotide in enumerate(seq1):\n",
    "                        if nukleotide not in nucleotides:\n",
    "                            last_nt_intron_index = i+1\n",
    "                        else:\n",
    "                            break\n",
    "                            \n",
    "                    id.append(alignment[0].id)\n",
    "                    exon_or_intron_number.append(index)\n",
    "                    is_it_intron.append(True) \n",
    "                    percent_of_homology_for_exons.append(None)\n",
    "                    length.append(last_nt_intron_index)\n",
    "                    first_10_nt.append(str(seq2[:10]))\n",
    "                    last_10_nt.append(str(seq2[last_nt_intron_index-10 : last_nt_intron_index]))\n",
    "                    sequence.append(str(seq2[:last_nt_intron_index]))\n",
    "                    path.append(str(file))\n",
    "                    class_of_exon.append(None)\n",
    "        \n",
    "                    seq1 = seq1[last_nt_intron_index:]\n",
    "                    seq2 = seq2[last_nt_intron_index:]\n",
    "                    \n",
    "                else:\n",
    "                    break\n",
    "                        \n",
    "            dictionary = {\"ID\":id,\n",
    "                          \"exon_or_intron_number\":exon_or_intron_number, \n",
    "                          \"is_it_intron\":is_it_intron, \n",
    "                          \"class_of_exon\":class_of_exon, \n",
    "                          \"percent_of_homology_for_exons\":percent_of_homology_for_exons, \n",
    "                          \"length\":length,\n",
    "                          \"first_10_nt\" : first_10_nt,\n",
    "                          \"last_10_nt\" : last_10_nt, \n",
    "                          \"path\":path, \n",
    "                          \"sequence\" : sequence}\n",
    "            data_frame = pd.DataFrame(dictionary)\n",
    "            #print(data_frame)\n",
    "        data_frames.append(data_frame)\n",
    "        #print(data_frames)\n",
    "\n",
    "\n",
    "    final_data_frame = pd.concat(data_frames, ignore_index=True)\n",
    "    final_data_frame.index = np.arange(1, len(final_data_frame) + 1)\n",
    "    print(final_data_frame)\n",
    "\n",
    "    def save_to_file(final_data_frame, filename):\n",
    "        if os.path.isfile(filename):\n",
    "            user_input = input(\"File already exists. Do you want overwrite? y/n: \")\n",
    "            if user_input == \"n\":\n",
    "                base_name, ext = os.path.splitext(filename)\n",
    "                i = 1\n",
    "                while os.path.isfile(f\"{base_name}_{i}{ext}\"):\n",
    "                    i += 1\n",
    "                filename = f\"{base_name}_{i}{ext}\"\n",
    "                final_data_frame.to_csv(filename, sep = \"\\t\")\n",
    "                \n",
    "            elif user_input == \"y\":\n",
    "                final_data_frame.to_csv(filename, sep = \"\\t\")\n",
    "            else:\n",
    "                print(\"Clarify your answer\")\n",
    "                \n",
    "        else:\n",
    "            final_data_frame.to_csv(filename, sep = \"\\t\")\n",
    "            \n",
    "\n",
    "    save_to_file(final_data_frame, \"przyciete_introny.tsv\")\n",
    "\n",
    "cutting_scrap_right('./alingment_gen_transkrypt/', 2)\n",
    "\n",
    "\n",
    "#zapisac w formacie gff\n",
    " #cele na teraz:\n",
    "\n",
    "#mamy dwie opcje, albo rÄcznie ustawiamy to dla gff przez np. tworzenie list, albo szukamy informacji jak obecne dane i ich rozszerzenia mogÄ posÅuÅ¼yÄ do konwertowania na formaty bardziej przyjazne gff. Ma to wiÄkszy prÃ³g wejÅcia ale wiÄcej uczy i minimalizuje bÅÄdy\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26b8e65e-efab-49a9-ad98-5a333d0d8065",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cee6a147-39e9-434a-8a4f-1fc300a377a9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
