{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "f037dd59-8d78-4bbc-b1f8-f878afe1f342",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import AlignIO, SeqIO, Seq, pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "e7f5ae9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "min_length_aligned_sequence = 30 #Minimal lenght of sequence which could be an exon\n",
    "extreme_homology = 0.97 #percentage of homology of sequence, treshold #I assume two faulty aligned nucleotides per 100 (98%) and one more nt because sometimes latest nt can move from end of one sequence to beginning next sequence\n",
    "middle_homology = 0.9 #used for classification of internal exons\n",
    "path_to_file_before_MAFFT = \"/home/norbert/mrdn/euglena/kod_i_pliki/surowe_pliki_plus_minus_500\"\n",
    "path_to_file_after_MAFFT = '/home/norbert/mrdn/euglena/kod_i_pliki/alingment_gen_transkrypt/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "038e46c9-9a33-498a-a7fc-618d1fa05a39",
   "metadata": {},
   "outputs": [],
   "source": [
    "#MEGA ISTOTNE PYTANIE - CO WYŚWIETLAMY????\n",
    "                    #CZY W KOLUMNIE SEQUENCE WYSWIETLAMY SEKWENCJE TRANKRYPTU CZY GENU?\n",
    "#######################################################################################\n",
    "#######################################################################################\n",
    "\n",
    "# 1.  Z jednego pliku fasta tworyzmy 3 pliki kazdy dla innego gatunku #FUNKCJA NAPISANA - CZEKAM NA ZWOLNIENIE SIE SERWERA\n",
    "# 2. robimy im maffta na zadanych parametrach #SKRYPT JEST NAPISANY\n",
    "# 2.5 ZMIENIAMY troche ta funkcje by byla zaadaptowana do nowego formatu plikow. Moze nawet fasta - sprawdzmy to\n",
    "# 3. Wklepujemy te pliki do naszej funkcji cutting scrap\n",
    "# 4. gff tabelka i wysylamy Ani\n",
    "# 5. analiza tych plikow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "e6d4c5ee-799f-427e-839a-f95917b6a430",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " \n",
      " EL_OG0030308_t2_g2_EL_PAM1_GP153_OV05 sequence is too short, sequence has not exons or other fault. \n",
      "\n",
      "                                        ID exon_or_intron_number is_it_intron  \\\n",
      "1    EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                    E1          0.0   \n",
      "2    EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                    I1          1.0   \n",
      "3    EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                    E2          0.0   \n",
      "4    EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                    I2          1.0   \n",
      "5    EG_OG0030307_t1_g1_EG_PAM1_GP153_OV05                    E3          0.0   \n",
      "..                                     ...                   ...          ...   \n",
      "144  EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                    E6          0.0   \n",
      "145  EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                    I6          1.0   \n",
      "146  EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                    E7          0.0   \n",
      "147  EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                    I7          1.0   \n",
      "148  EG_OG0030309_t3_g3_EG_PAM1_GP153_OV05                    E8          0.0   \n",
      "\n",
      "    class_of_exon  percent_of_homology_for_exons  length  first_nt_position  \\\n",
      "1               1                          97.92    96.0             2462.0   \n",
      "2            None                            NaN   102.0             2463.0   \n",
      "3               1                         100.00   145.0             2709.0   \n",
      "4            None                            NaN   238.0             2710.0   \n",
      "5               1                         100.00    72.0             3019.0   \n",
      "..            ...                            ...     ...                ...   \n",
      "144             1                          99.35   154.0             5586.0   \n",
      "145          None                            NaN   356.0             5587.0   \n",
      "146             1                         100.00    84.0             6026.0   \n",
      "147          None                            NaN  1669.0             6027.0   \n",
      "148             1                         100.00    67.0             7762.0   \n",
      "\n",
      "     last_nt_position first_10_nt  last_10_nt  \\\n",
      "1              2367.0  ctacacgtac  ggctgctcct   \n",
      "2              2564.0  gcccgccgcc  cggcgggggc   \n",
      "3              2565.0  gatcgagctg  ccgctgcttt   \n",
      "4              2947.0  gaccgggcgg  gcccgcggtc   \n",
      "5              2948.0  atgtcgccgg  caccgtggca   \n",
      "..                ...         ...         ...   \n",
      "144            5433.0  ctgagatcgt  tgattggcca   \n",
      "145            5942.0  gcgcccagtg  gtcactgcgg   \n",
      "146            5943.0  gcaattgcgg  tggattctga   \n",
      "147            7695.0  gcgccgaagt  gcgcttgcac   \n",
      "148            7696.0  gcatctgatg  tgtatgccct   \n",
      "\n",
      "                                                  path  \\\n",
      "1    /home/norbert/mrdn/euglena/kod_i_pliki/alingme...   \n",
      "2    /home/norbert/mrdn/euglena/kod_i_pliki/alingme...   \n",
      "3    /home/norbert/mrdn/euglena/kod_i_pliki/alingme...   \n",
      "4    /home/norbert/mrdn/euglena/kod_i_pliki/alingme...   \n",
      "5    /home/norbert/mrdn/euglena/kod_i_pliki/alingme...   \n",
      "..                                                 ...   \n",
      "144  /home/norbert/mrdn/euglena/kod_i_pliki/alingme...   \n",
      "145  /home/norbert/mrdn/euglena/kod_i_pliki/alingme...   \n",
      "146  /home/norbert/mrdn/euglena/kod_i_pliki/alingme...   \n",
      "147  /home/norbert/mrdn/euglena/kod_i_pliki/alingme...   \n",
      "148  /home/norbert/mrdn/euglena/kod_i_pliki/alingme...   \n",
      "\n",
      "                                              sequence  \n",
      "1    ccacaggtacaagcacgtgaacgtcatggcgccggaggtccaaagc...  \n",
      "2    gcccgccgcccggggttcggtgcccccggctgtcaggccgttgggt...  \n",
      "3    gatcgagctgctgaccggccagccggcgttccccgtctcggaggac...  \n",
      "4    gaccgggcggcctcaccatggcagtgccacatcctgggccaacaga...  \n",
      "5    atgtcgccggacgccctctcgctggccctccgctgcctcgccaagt...  \n",
      "..                                                 ...  \n",
      "144  ctgagatcgtcttccagggctgccagcagctccgcacgcagctcca...  \n",
      "145  gcgcccagtgccttgcggggcccactggcagtgctgcccaggaacc...  \n",
      "146  gcaattgcggctccatgcgcagtacgtgcggggggagagccacgcg...  \n",
      "147  gcgccgaagtgccatttgttggaacctgaggattttccttaggggt...  \n",
      "148  gcatctgatgaaggcgtccatagcaaccgccaaacggatggccctc...  \n",
      "\n",
      "[148 rows x 12 columns]\n",
      "Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "def cutting_scrap(path_to_file_before_MAFFT, path_to_file_after_MAFFT, acceptable_gap_length):\n",
    "    data_frames = []\n",
    "    gff_data_frames = []\n",
    "    gaps_signs = \"-\" * acceptable_gap_length #maximum length of gaps in sequence in one exon's sequence\n",
    "    \n",
    "    def extracting_strands_from_alignment(alignment): #that function describe what strand is transcript strand and genomic strand\n",
    "        upper_seq = alignment[0].seq\n",
    "        lower_seq = alignment[1].seq\n",
    "        gaps_amount_upper = upper_seq.count(\"-\")\n",
    "        gaps_ammount_lower = lower_seq.count(\"-\")\n",
    "        if gaps_amount_upper > gaps_ammount_lower:\n",
    "            seq1 = upper_seq #transcript sequence\n",
    "            seq2 = lower_seq #gene sequence\n",
    "        else:\n",
    "            seq1 = lower_seq\n",
    "            seq2 = upper_seq\n",
    "        return seq1, seq2\n",
    "\n",
    "    #def setting_first_nucleotide_pair(alignment): #ta funkcja ucina nam 6 odczytow #to nam nie działa więc nie brnę dalej. Nie działa prawdopodobnie dlatego, że jest ona próba odwzorowania kodu, który spokojnie sobie działał w pętli while true\n",
    "        #if \"-\" in alignment[:, 0]: #deleting mismatches at the first nucleotide\n",
    "            #return alignment[:, 1:]\n",
    "        #else:\n",
    "            #return alignment\n",
    "            #print(f\"alignment after cut:\\n{alignment}\") #chekpoint\n",
    "\n",
    "\n",
    "    #def setting_last_nuckleotide_pair(alignment): #to nam nie działa więc nie brnę dalej\n",
    "        #if \"-\" in alignment[:, -1]: #deleting mismatches at the end\n",
    "            #alignment = alignment[:, :-1]\n",
    "            #return alignment \n",
    "        #else:\n",
    "            #return alignment #print(f\"alignment check:\\n{alignment[:, -50:-1]}\") checkpoint\n",
    "                    \n",
    "\n",
    "    def finding_gap_index(sequence, acceptable_gap_length):\n",
    "        return sequence.find(\"-\" * acceptable_gap_length) \n",
    "\n",
    "\n",
    "    def reversed_finding_gap_index(sequence, acceptable_gap_length):\n",
    "        return sequence.rfind(\"-\" * acceptable_gap_length) + acceptable_gap_length #now we have to add length of gap to index, because python starts counting from 0 nt, and we just change direction of counting.\n",
    "\n",
    "    def setting_distance_between_first_nt_and_gap(is_it_exon):\n",
    "        if is_it_exon == True:\n",
    "            if first_gap_position_seq2 < 0: #when in second strand there are no gaps - it means, second strand is correctly prepared\n",
    "                distance_between_first_nt_and_gap = first_gap_position_seq1\n",
    "            else:\n",
    "                distance_between_first_nt_and_gap = min(first_gap_position_seq1, first_gap_position_seq2)\n",
    "            return distance_between_first_nt_and_gap\n",
    "        elif is_it_exon == False:\n",
    "            if last_gap_position_seq2 < 0: #it means, if there is no gaps in second sequention, gap position is gap position in seq1. Additional, rfind() function return \"-1\" if it won't find any gaps in seq\n",
    "                distance_between_first_nt_and_gap = (len(seq1) - last_gap_position_seq1) -1\n",
    "                #print(\"pierwszy\", distance_between_first_nt_and_gap) #checkpoint\n",
    "            else:\n",
    "                distance_between_first_nt_and_gap = (len(seq1) - max(last_gap_position_seq1, last_gap_position_seq2))\n",
    "            return distance_between_first_nt_and_gap\n",
    "                #print(\"drugi\", distance_between_first_nt_and_gap) #checkpoint\n",
    "\n",
    "            \n",
    "    def describing_class_of_exon(percent_of_homology, last_nt_exon_index, min_length_aligned_sequence, class_of_exon):\n",
    "        if percent_of_homology >= 0.9 and last_nt_exon_index >= min_length_aligned_sequence :\n",
    "            class_of_exon.append(\"1\")\n",
    "        elif percent_of_homology < 0.9 and percent_of_homology > 0 and last_nt_exon_index < min_length_aligned_sequence:\n",
    "            class_of_exon.append(\"2\")\n",
    "        else:\n",
    "            class_of_exon.append(\"3\")\n",
    "\n",
    "  \n",
    "    def exons_indices(is_it_exon, max_genomic_seq, seq2, last_nt_exon_index, first_nt_position, last_nt_position):\n",
    "        if is_it_exon:\n",
    "            first_nt_index_of_exon = max_genomic_seq.find(str(seq2[:last_nt_exon_index]).replace(\"-\", \"\"))\n",
    "            if first_nt_index_of_exon == -1:\n",
    "                first_nt_position.append(0)\n",
    "                last_nt_position.append(0)\n",
    "                print(f\"That exon: {seq2[:last_nt_exon_index].replace('-', '')} comes from genomic sequence and cannot be found in file_before_MAFFT.\")\n",
    "            else:\n",
    "                first_nt_position.append(first_nt_index_of_exon + 1)\n",
    "                last_nt_position.append(first_nt_index_of_exon + len(seq2[:last_nt_exon_index]))\n",
    "                \n",
    "\n",
    "\n",
    "    def introns_indices(is_it_exon, max_genomic_seq, seq2, last_nt_intron_index, first_nt_position, last_nt_position):\n",
    "        if is_it_exon: \n",
    "            first_nt_index_of_intron = max_genomic_seq.find(str(seq2[:last_nt_intron_index]))\n",
    "            if first_nt_index_of_intron == -1:\n",
    "                first_nt_position.append(0)\n",
    "                last_nt_position.append(0)\n",
    "            else:\n",
    "                first_nt_position.append(first_nt_index_of_intron + 1)\n",
    "                last_nt_position.append(first_nt_index_of_intron + len(seq2[:last_nt_intron_index]))\n",
    "                \n",
    "\n",
    "    \n",
    "    def save_to_file(final_data_frame, filename):\n",
    "        if os.path.isfile(filename):\n",
    "            user_input = input(\"File already exists. Do you want overwrite? y/n: \")\n",
    "            if user_input == \"n\":\n",
    "                base_name, ext = os.path.splitext(filename)\n",
    "                i = 1\n",
    "                while os.path.isfile(f\"{base_name}_{i}{ext}\"):\n",
    "                    i += 1\n",
    "                filename = f\"{base_name}_{i}{ext}\"\n",
    "                final_data_frame.to_csv(filename, sep = \"\\t\")\n",
    "                \n",
    "            elif user_input == \"y\":\n",
    "                final_data_frame.to_csv(filename, sep = \"\\t\")\n",
    "            else:\n",
    "                print(\"Clarify your answer. Nothing has done.\")\n",
    "                \n",
    "        else:\n",
    "            final_data_frame.to_csv(filename, sep = \"\\t\")\n",
    "            \n",
    "    \n",
    "    for filename in os.listdir(path_to_file_after_MAFFT):\n",
    "        file = os.path.join(path_to_file_after_MAFFT, filename)\n",
    "        if not os.path.isfile(file):\n",
    "            continue\n",
    "        if file.endswith(\".aln\"):\n",
    "            alignment = AlignIO.read(file, \"clustal\")\n",
    "            \n",
    "            #czy ten fragment trzeba zamknąć do funkcji i wyeksportować ją na początek - mówie tu o kwestiach estetycznych\n",
    "            for filename_before_MAFFT in os.listdir(path_to_file_before_MAFFT):\n",
    "                file_before_MAFFT = os.path.join(path_to_file_before_MAFFT, filename_before_MAFFT)\n",
    "                if not os.path.isfile(file_before_MAFFT):\n",
    "                    #print(f\"This thingy: {str(filename_before_MAFFT)} could not be used to pair with alignment. Check it or ignore\")\n",
    "                    continue\n",
    "                if filename_before_MAFFT.endswith(\".fasta\"):\n",
    "                    if filename.startswith(filename_before_MAFFT[:12]):\n",
    "                        max_len = 0\n",
    "                        max_genomic_seq = \"\"\n",
    "                        for record in SeqIO.parse(file_before_MAFFT, \"fasta\"): #getting genomic sequence - usually it is longer seq than read\n",
    "                            sequence = record.seq\n",
    "                            if len(sequence) > max_len:\n",
    "                                max_len = len(sequence)\n",
    "                                max_genomic_seq = record.seq   \n",
    "                         \n",
    "            while True:\n",
    "                is_it_exon = True\n",
    "                #alignment = setting_first_nucleotide_pair(alignment) #function that doesn't work\n",
    "                first_nucleotides_pair = alignment[:, 0]#first nucleotides in both strands as variable\n",
    "                if \"-\" in first_nucleotides_pair: #deleting mismatches at the start \n",
    "                    alignment = alignment[:, 1:]\n",
    "                    #print(f\"alignment after cut:\\n{alignment}\") #chekpoint\n",
    "                    continue\n",
    "    \n",
    "                \n",
    "                seq1, seq2 = extracting_strands_from_alignment(alignment)\n",
    "\n",
    "                \n",
    "                first_gap_position_seq1 = finding_gap_index(seq1, acceptable_gap_length)\n",
    "                first_gap_position_seq2 = finding_gap_index(seq2, acceptable_gap_length)\n",
    "                #print(f\"first gap position in first sequence:{first_gap_position_seq1}, in 2nd sequence: {first_gap_position_seq2}\") #checkpoint\n",
    "\n",
    "                distance_between_first_nt_and_gap = setting_distance_between_first_nt_and_gap(is_it_exon)\n",
    "                \n",
    "                local_alignment = pairwise2.align.localxx(seq1[:distance_between_first_nt_and_gap], seq2[:distance_between_first_nt_and_gap], one_alignment_only = True) #\"xx\"  means no gap penalty while opening gaps or longering them and no penalties for mismatch. Just pure score of alignment to count homology\n",
    "                #print(\"local_alignment:\", local_alignment) #checkpoint\n",
    "                   \n",
    "                try:  #instruction what to do when seq is too short\n",
    "                    local_homology_percentage = (local_alignment[0].score / distance_between_first_nt_and_gap) \n",
    "                except IndexError:\n",
    "                    if len(seq1) < 100:\n",
    "                        print(f\"\\n \\n {alignment[0].id} sequence is too short, sequence has not exons or other fault. \\n\")\n",
    "                        break\n",
    "                    continue\n",
    "                    \n",
    "\n",
    "                \n",
    "                if local_homology_percentage <= extreme_homology:\n",
    "                    #print(f\"\\n Too low score of homology. \\n alignment: {alignment[:, :50]} \\n score: {local_alignment[0].score}, distance: {distance_between_first_nt_and_gap}, percent: {round(local_homology_percentage, 2)*100}%, \\n {local_alignment}\")\n",
    "                    alignment = alignment[:, distance_between_first_nt_and_gap:] \n",
    "                    \n",
    "                else:\n",
    "                    if distance_between_first_nt_and_gap <= min_length_aligned_sequence:\n",
    "                        #print(f\"\\n alignment is too short. \\n distance between sequences {distance_between_first_nt_and_gap}, \\n alignment: {alignment}, \\n score: {local_alignment[0].score}, distance: {distance_between_first_nt_and_gap}, procentowo: {round(local_homology_percentage, 2)*100}%, \\n {local_alignment}\")\n",
    "                        alignment = alignment[:, distance_between_first_nt_and_gap:]\n",
    "                    else:\n",
    "                        #print(f\"Sequence {alignment[0].id} cut properly from LEFT side, \\n distance between sequences: {distance_between_first_nt_and_gap}\"), alignment: {alignment[:, :50]},\n",
    "                        break\n",
    "    \n",
    "               \n",
    "        \n",
    "            \n",
    "        ######################### TERAZ OD PRAWEJ DO LEWEJ #######################  \n",
    "\n",
    "    \n",
    "            \n",
    "            while True:\n",
    "                is_it_exon = False\n",
    "                last_nucleotides_pair = alignment[:, -1]#last nucleotides in both strands as variable\n",
    "                if \"-\" in last_nucleotides_pair: #deleting mismatches at the end\n",
    "                    alignment = alignment[:, :-1]\n",
    "                     #print(f\"alignment check:\\n{alignment[:, -50:-1]}\") checkpoint\n",
    "                    continue\n",
    "                             \n",
    "                seq1, seq2 = extracting_strands_from_alignment(alignment)\n",
    "\n",
    "                last_gap_position_seq1 = reversed_finding_gap_index(seq1, acceptable_gap_length)\n",
    "                last_gap_position_seq2 = reversed_finding_gap_index(seq2, acceptable_gap_length)\n",
    "                #print(\"seq1:\", last_gap_position_seq1, \"seq2:\", last_gap_position_seq2) #checkpoint\n",
    "                \n",
    "                distance_between_first_nt_and_gap = setting_distance_between_first_nt_and_gap(is_it_exon)\n",
    "                #print(\"drugi\", distance_between_first_nt_and_gap)\n",
    "                        \n",
    "                local_alignment = pairwise2.align.localxx(seq1[-distance_between_first_nt_and_gap:], seq2[-distance_between_first_nt_and_gap:], one_alignment_only = True) #\"xx\" means no gap penalty while opening gaps or longering them and no penalties for mismatch. Just pure score of alignment\n",
    "                #print(\"local_alignment:\", local_alignment) #checkpoint\n",
    "                \n",
    "                \n",
    "        \n",
    "                try:#instruction what to do when seq is too short or gaps were not found\n",
    "                    if distance_between_first_nt_and_gap != 0:\n",
    "                        local_homology_percentage = (local_alignment[0].score / distance_between_first_nt_and_gap) \n",
    "                    else:\n",
    "                        break\n",
    "                except IndexError:\n",
    "                    if len(seq1) < 100:\n",
    "                        print(f\"\\n \\n {alignment[0].id} sequence is too short or sequence has not exons. \\n\")\n",
    "                        break\n",
    "                    continue\n",
    "        \n",
    "                \n",
    "                if local_homology_percentage <= extreme_homology:\n",
    "                     #print(f\"\\n Too low homology! \\n alignment: {alignment[:, -50:]} \\n score: {local_alignment[0].score}, \\distance: {distance_between_first_nt_and_gap}, %%%: {round(local_homology_percentage, 2)*100}%, \\n {local_alignment}\")\n",
    "                    alignment = alignment[:, :-distance_between_first_nt_and_gap] \n",
    "                    \n",
    "                else:\n",
    "                    if distance_between_first_nt_and_gap <= min_length_aligned_sequence:\n",
    "                        #print(f\"\\n Alignment is too short!. \\n distance: {distance_between_first_nt_and_gap}, \\n alignment: {alignment}, \\n score: {local_alignment[0].score}, distance: {distance_between_first_nt_and_gap}, %%%: {round(local_homology_percentage, 2)*100}%, \\n {local_alignment}\")\n",
    "                        alignment = alignment[:, :-distance_between_first_nt_and_gap]\n",
    "                    else:\n",
    "                        #print(f\"Sequence: {alignment[0].id} cut properly from RIGHT side, alignment: \\n {alignment[:, -50:]}\")\n",
    "                        break\n",
    "\n",
    "        \n",
    "        \n",
    "            \n",
    "            nucleotides = [\"a\", \"t\", \"g\", \"c\", \"A\", \"T\", \"G\", \"C\"]\n",
    "            index = 0\n",
    "            \n",
    "            id = []\n",
    "            exon_or_intron_number = []\n",
    "            is_it_intron = []\n",
    "            class_of_exon = []\n",
    "            percent_of_homology_for_exons = []\n",
    "            length = []\n",
    "            first_nt_position = []\n",
    "            last_nt_position = []\n",
    "            first_10_nt = []\n",
    "            last_10_nt = []\n",
    "            path = []\n",
    "            sequence = []\n",
    "\n",
    "            \n",
    "            while True: #instruction what to do when seq is too short or gaps were not found\n",
    "                try: \n",
    "                    if distance_between_first_nt_and_gap != 0:\n",
    "                        local_homology_percentage = (local_alignment[0].score / distance_between_first_nt_and_gap) #that sign \"[]\" becouse local_alignment's type is list\n",
    "                    else:\n",
    "                        break\n",
    "                except IndexError:\n",
    "                    if len(seq1) < 100:\n",
    "                        print(f\"\\n \\n {alignment[0].id} sequence is too short or sequence has not exons. \\n\")\n",
    "                        break\n",
    "                    continue\n",
    "\n",
    "                if len(seq1) > 1 and seq1[0] != gaps_signs:#mining data from exons\n",
    "                    is_it_exon = True\n",
    "                    index += 1\n",
    "                    last_nt_exon_index = seq1.find(gaps_signs) \n",
    "                \n",
    "                    if last_nt_exon_index == -1: #it means if it is last exon, if find() don't find any gap signs at the end\n",
    "                        last_nt_exon_index = len(seq1)\n",
    "                                \n",
    "                    alignment_of_exon = pairwise2.align.localxx(seq1[:last_nt_exon_index], seq2[:last_nt_exon_index], one_alignment_only = True)\n",
    "                        #print( \"\\n alignment since start to the end of exon:\", alignment_of_exon[0])\n",
    "                    percent_of_homology = alignment_of_exon[0].score / last_nt_exon_index\n",
    "                        #print(\"exon analising finished\")\n",
    "                            \n",
    "            \n",
    "                    #extending lists to make bigger dictionary and then data frame\n",
    "                    id.append(alignment[0].id)\n",
    "                    exon_or_intron_number.append(f\"E{index}\")\n",
    "                    is_it_intron.append(False)\n",
    "                    percent_of_homology_for_exons.append(round(percent_of_homology * 100, 2))\n",
    "                    length.append(last_nt_exon_index)\n",
    "                    first_10_nt.append(str(seq1[:10]))\n",
    "                    last_10_nt.append(str(seq2[last_nt_exon_index-10 : last_nt_exon_index])) \n",
    "                    sequence.append(str(seq2[:last_nt_exon_index]).replace(\"-\", \"\"))\n",
    "                    path.append(str(file))\n",
    "       \n",
    "                    exons_indices(is_it_exon, max_genomic_seq, seq2, last_nt_exon_index, last_nt_position, first_nt_position)\n",
    "                    \n",
    "                    describing_class_of_exon(percent_of_homology, last_nt_exon_index, min_length_aligned_sequence, class_of_exon)\n",
    "            \n",
    "                    seq1 = seq1[last_nt_exon_index:]\n",
    "                    seq2 = seq2[last_nt_exon_index:]\n",
    "                    #print(\"analising exon finished\") #checkpoint\n",
    "                            \n",
    "                #expanding index\n",
    "                if len(seq1) > 1 and seq1[0] == \"-\":\n",
    "                    is_is_exon = False\n",
    "                    for i, nukleotide in enumerate(seq1):\n",
    "                        if nukleotide not in nucleotides:\n",
    "                            last_nt_intron_index = i+1\n",
    "                        else:\n",
    "                            break\n",
    "\n",
    "                    #extending lists to make bigger dictionary and then data frame\n",
    "                    id.append(alignment[0].id)\n",
    "                    exon_or_intron_number.append(f\"I{index}\")\n",
    "                    is_it_intron.append(True)\n",
    "                    class_of_exon.append(None)\n",
    "                    percent_of_homology_for_exons.append(None)\n",
    "                    length.append(last_nt_intron_index)\n",
    "                    #first_nt_position.append(0)\n",
    "                    #last_nt_position.append(0)\n",
    "                    first_10_nt.append(str(seq2[:10]))\n",
    "                    last_10_nt.append(str(seq2[last_nt_intron_index-10 : last_nt_intron_index]))\n",
    "                    sequence.append(str(seq2[:last_nt_intron_index]).replace(\"-\", \"\"))\n",
    "                    path.append(str(file))\n",
    "\n",
    "                    introns_indices(is_it_exon, max_genomic_seq, seq2, last_nt_intron_index, first_nt_position, last_nt_position)\n",
    "\n",
    "                    seq1 = seq1[last_nt_intron_index:]\n",
    "                    seq2 = seq2[last_nt_intron_index:]\n",
    "                    \n",
    "                else:\n",
    "                     break\n",
    "                        \n",
    "            dictionary = {\"ID\":id,\n",
    "                          \"exon_or_intron_number\":exon_or_intron_number, \n",
    "                          \"is_it_intron\":is_it_intron, \n",
    "                          \"class_of_exon\":class_of_exon, \n",
    "                          \"percent_of_homology_for_exons\":percent_of_homology_for_exons, \n",
    "                          \"length\":length,\n",
    "                          \"first_nt_position\" : first_nt_position,\n",
    "                          \"last_nt_position\" : last_nt_position,\n",
    "                          \"first_10_nt\" : first_10_nt,\n",
    "                          \"last_10_nt\" : last_10_nt, \n",
    "                          \"path\":path, \n",
    "                          \"sequence\" : sequence}\n",
    "#EL|STRG.5113.1|scaffold_3161:4719-42082 manual_annotation   transcript  505 37862   1000    +   .   EL|STRG.5113.1|scaffold_3161:4719-42082 #tak ma wygladac calosc\n",
    "#EL STRG.5113.1       nazwa backbone\n",
    "#scaffold_3161:4719-42082 manual_annotation            źródło (np stringtie, manual)\n",
    "#transcript           transcript/exon\n",
    "#505        start\n",
    "#37862      end\n",
    "#1000       pokrycie\n",
    "#+          strand (+ lub -)\n",
    "#.          kropka nie wiadomo po co\n",
    "#EL|STRG.5113.1|scaffold_3161:4719-42082          dłuższa nazwa sekwencji (gene_id \"STRG.1\"; transcript_id \"STRG.1.1\"; cov \"52.968773\"; FPKM \"5.233016\"; TPM \"3.238127\";) #jakies tam inne informacje oddzielone tym czyms ;;;;\n",
    "\n",
    "\n",
    "\n",
    "#OG0031981.fasta:LON SL- GGOE01042269.1\n",
    "#OG0031981.fasta:GRA SL- TRINITY_DN37780_c0_g2_i5\n",
    "#OG0031981.fasta:HIE SL- TRINITY_DN53499_c0_g1_i4\n",
    "#OG0031981.fasta:LON scaffold_16993\n",
    "#OG0031981.fasta:GRA Backbone_7059\n",
    "#OG0031981.fasta:HIE scaffold_10332\n",
    "#OG0031981 + HIE/LON/GRA musi byc seqid\n",
    "#TRINITY scaffold Backbone + numerki na końcu to\n",
    "#+ - strand\n",
    "#dodatkowe informacje o pokryciu a moze tez tutaj bedzie ID\n",
    "\n",
    "  \n",
    "            seqid = [] #name of the chromosome or scaffold; chromosome names can be given with or without the 'chr' prefix. Important note: the seq ID must be one used within Ensembl, i.e. a standard chromosome name or an Ensembl identifier such as a scaffold ID, without any additional content such as species or assembly. See the example GFF output below.\n",
    "            source = [] #tutaj pominiemy ten krok - damy kropke #Describes the algorithm or the procedure that generated this feature. Typically Genescane or Genebank, respectively. #scaffold backbone trinity  i te numerki jakieś\n",
    "            type = [] #Describes what the feature is (mRNA, domain, exon, etc.). #scaffold backbone trinity  i te numerki jakieś #intron lub exon\n",
    "            start = []\n",
    "            end = []\n",
    "            score = [] #Typically E-values for sequence similarity and P-values for predictions. Homology there\n",
    "            strand = []\n",
    "            phase = [] #Indicates where the feature begins with reference to the reading frame. The phase is one of the integers 0, 1, or 2, indicating the number of bases that should be removed from the beginning of this feature to reach the first base of the next codon.. Propably 0\n",
    "            attributes = [] #A semicolon-separated list of tag-value pairs, providing additional information about each feature. Some of these tags are predefined, e.g. ID, Name, Alias, Parent . You can see the full list [here](https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md).\n",
    "            data_frame = pd.DataFrame(dictionary)\n",
    "\n",
    "            \n",
    "\n",
    "\n",
    "\n",
    "            \n",
    "            gff_dictionary = {}\n",
    "\n",
    "            \n",
    "            gff_data_frame = pd.DataFrame(gff_dictionary)\n",
    "            #print(data_frame)\n",
    "        data_frames.append(data_frame)\n",
    "        gff_data_frames.append(gff_data_frame)\n",
    "        #print(data_frames)\n",
    "\n",
    "\n",
    "    final_data_frame = pd.concat(data_frames, ignore_index=True)\n",
    "    final_data_frame.index = np.arange(1, len(final_data_frame) + 1)\n",
    "    print(final_data_frame)\n",
    "    \n",
    "    gff_final_data_frame = pd.concat(gff_data_frames, ignore_index = True)\n",
    "    gff_final_data_frame.index = np.arange(1, len(gff_final_data_frame) + 1)\n",
    "    print(gff_final_data_frame)\n",
    "    \n",
    "    #save_to_file(final_data_frame, \"przyciete_introny.tsv\")\n",
    "\n",
    "\n",
    "cutting_scrap(path_to_file_before_MAFFT, path_to_file_after_MAFFT, 2)\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2a04e46-1593-4d3e-9423-54efbcfe4bbe",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
